<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>SARS-CoV-2 Genomics</title>

<script src="site_libs/header-attrs-2.16/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="assets/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">UoC BioinfoTraining</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="101-setup.html">
    <span class="fa fa-gear"></span>
     
    Setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-file-lines"></span>
     
    Materials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">The Unix Shell</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="02a-unix_intro.html">Intro to Unix</a>
        </li>
        <li>
          <a href="02b-unix_files_directories.html">Files and Directories</a>
        </li>
        <li>
          <a href="02c-unix_text_manipulation.html">Text Manipulation</a>
        </li>
        <li>
          <a href="02d-unix_pipes.html">Combining Commands &amp; Writing Scripts</a>
        </li>
      </ul>
    </li>
    <li>
      <a href="01-intro.html">SARS-CoV-2 Genomic Surveillance</a>
    </li>
    <li>
      <a href="03-intro_ngs.html">Intro to NGS Data</a>
    </li>
    <li>
      <a href="04-consensus.html">Consensus Assembly</a>
    </li>
    <li>
      <a href="05-lineage_analysis.html">Lineages and Variants</a>
    </li>
    <li>
      <a href="06-phylogeny.html">Building Phylogenies</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Extras
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="102-unix_cheatsheet.html">Unix Cheatsheet</a>
    </li>
    <li>
      <a href="106-file_formats.html">Common File Formats</a>
    </li>
    <li>
      <a href="104-wsl_windows.html">Windows Subsystem for Linux</a>
    </li>
    <li>
      <a href="105-vs_code.html">Visual Studio Code</a>
    </li>
    <li>
      <a href="103-tools_and_resources.html">SARS-CoV-2 Resources</a>
    </li>
  </ul>
</li>
<li>
  <a href="100-homework.html">
    <span class="fa fa-pencil"></span>
     
    Homework
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cambiotraining/sars-cov-2-genomics">
    <span class="fa fa-brands fa-github"></span>
     
    Github
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<div class="warning">
<p><strong>Update May 2022</strong></p>
<p>Please note that this is an old version of this lesson for workshops that ran prior to May 2022. We kept this in the materials as a reference, but would recommend that you look at the <a href="05-lineage_analysis.html">new version of the lesson</a> instead.</p>
</div>
<div id="lineage-assignment-and-variant-classification" class="section level1">
<h1>Lineage Assignment and Variant Classification</h1>
<div class="highlight">
<p><strong>Questions</strong></p>
<ul>
<li>What are the main variant annotation conventions for SARS-CoV-2?</li>
<li>How can I assign sequences to lineages and classify them as variants of interest/concern?</li>
<li>How can I visually explore the results of my variant analysis?</li>
</ul>
<p><strong>Learning Objectives</strong></p>
<ul>
<li>Understand variant annotation conventions used by Gisaid, Pango, Nextstrain and WHO and how they relate to each other.</li>
<li>Assign sequences to Pango lineages using <code>pangolin</code>.</li>
<li>Interactively explore mutations in the assembled genomes and their phylogenetic context using <em>Nexstrain</em>’s tools.</li>
</ul>
</div>
<div class="note">
<p>This section has an accompanying <a href="https://docs.google.com/presentation/d/1urubrzjF6VzpWfknqnYMjbMcIQwcO2p1OQSE9-paAUI/edit?usp=sharing" target="_blank">slide deck</a>.</p>
</div>
<div id="sars-cov-2-variants" class="section level2">
<h2>SARS-CoV-2 Variants</h2>
<p>As viruses (or any other organism) evolve, random DNA changes occur in the population, for example due to replication errors. Many of these changes are likely to be <em>neutral</em>, meaning that they do not change the characteristics of the virus in any significant way. Neutral changes tend to <em>drift</em> in the population, increasing or decreasing in frequency in a random way, but most likely end up disappearing due to their low starting frequency in the population.</p>
<p>On the other hand, <em>advantageous mutations</em> can occur, which lead to changes in the characteristics of the virus that are beneficial for its spread in the population (e.g. high transmissibility, resistance to immune response or vaccines, etc.). Such beneficial mutations will therefore experience <em>positive selection</em>, potentially leading to their increase in frequency in the population as they spread to more and more hosts.</p>
<p>Viruses carrying those advantageous mutations may, over time, aquire other advantageous mutations that further increase their fitness and, therefore, their frequency in the population. One way to visualise this is by looking at a <em>phylogenetic tree</em> showing the relationship between sequences (based on their similarity) and how groups of sequences change over time.</p>
<div class="figure">
<img src="images/lineages_example.svg" alt="" />
<p class="caption">Example of global phylogeny from the <a href="https://nextstrain.org/ncov/gisaid/global">Nextstrain public server</a>. Colours show different Nextstrain clades. (Screenshot taken Feb 2022)</p>
</div>
<p>In the figure above, which shows SARS-CoV-2 samples from across the world, we can see groups of similar sequences rapidly “expanding” at certain points in time. Such groups of sequences, which share a collection of DNA changes, are referred to as SARS-CoV-2 <em>variants</em> (see box below about the ambiguous meaning of this term). In an effort to understand the spread of the virus and monitor situations of increased occurrence of such variants, several groups and institutions have developed a system to classify groups of SARS-CoV-2 sequences as <em>variants of interest</em> and <em>variants of concern</em>.</p>
<p>A full explanation and definitions of such variants is given in the <a href="https://www.who.int/en/activities/tracking-SARS-CoV-2-variants">World Health Organisation (WHO) variants page</a> The main classification systems currently in use are:</p>
<ul>
<li><a href="https://www.gisaid.org/references/statements-clarifications/clade-and-lineage-nomenclature-aids-in-genomic-epidemiology-of-active-hcov-19-viruses/">GISAID clades</a></li>
<li><a href="https://github.com/nextstrain/ncov/blob/master/defaults/clades.tsv">Nextstrain clades</a></li>
<li><a href="https://cov-lineages.org/lineage_list.html">Pango lineages</a></li>
<li><a href="https://www.who.int/en/activities/tracking-SARS-CoV-2-variants">World Health Organisation (WHO) variants</a></li>
</ul>
<p>In practice, there is a big overlap between these different nomenclature systems, with WHO variants having a direct match to Pango lineages and Nextstrain clades. In fact, the different teams work together to try and harmonise the nomenclature used, and thus facilitate the interpretation of sequence analysis.</p>
<p>The two most popular systems - <em>Nextclade</em> and <em>Pangolin</em> - have slightly different levels of resolution. <em>Nextclade</em>’s <a href="https://nextstrain.org/blog/2021-01-06-updated-SARS-CoV-2-clade-naming">nomenclature system</a> was developed to highlight diversity patterns at a larger scale, allowing discussions of SARS-CoV-2 diversity at a global level and over larger time scales. On the other hand, <em>Pangolin</em>’s <a href="https://www.pango.network/the-pango-nomenclature-system/statement-of-nomenclature-rules/">nomenclature system</a> is more fine-grained, aiming to follow the dynamics of the pandemic as it unfolds. The two systems are complementary to each other, and our analysis of SARS-CoV-2 sequences should include both tools.</p>
<div class="note">
<p><strong>What is a variant?</strong></p>
<p>It is important to note that the term <em>variant</em> can be sometimes ambiguous.</p>
<p>The term “<strong>SARS-CoV-2 variant</strong>” usually refers to the <a href="https://www.who.int/en/activities/tracking-SARS-CoV-2-variants/">WHO definition of variants of concern/interest</a> (e.g. the Alpha, Delta and Omicron variants), which includes sequences containing a <em>collection of several nucleotide changes</em> that characterise that group. According to this definition, we have two variants in the example below (samples 1 &amp; 2 are one variant and samples 3 &amp; 4 another variant).</p>
<p><img src="images/variants_snps_indels.svg" /></p>
<p>However, in bioinformatic sequence analysis, a <strong>sequence variant</strong> refers to an individual change in the DNA sequence (a SNP or an insertion/deletion). Using this definition, in the example above we have 5 variants: 3 SNPs and 2 indels. In the <a href="04-artic_nextflow.html">Consensus Sequence</a> section, we mentioned one of our workflow steps was “variant calling”. This was the definition of variant we were using: identifying individual SNPs and/or indels relative to the reference genome, from our sequencing data. This is also reflected in the one of the common file formats used to store SNP/indel information, the <a href="https://en.wikipedia.org/wiki/Variant_Call_Format"><em>VCF</em> file</a>, which means “Variant Call Format”.</p>
<p>Sometimes the term “mutation” is used to refer to SNP/indel variants. For example see this <a href="https://www.cogconsortium.uk/what-do-virologists-mean-by-mutation-variant-and-strain/">definition from the COG consortium</a>.</p>
<p>Because of this ambiguity, the terms “lineages” or “clades” are often used instead of “variants” when referring to groups of similar SARS-CoV-2 sequences, because they have a phylogenetic interpretation.</p>
</div>
</div>
<div id="pangolin" class="section level2">
<h2>Pangolin</h2>
<p><img src="https://cov-lineages.org/assets/images/pangolin_logo.svg" alt="Pangolin" style="float:right;width:20%"></p>
<p>The first tool we will use is called <code>pangolin</code> and uses the <a href="https://cov-lineages.org/">Pango nomenclature system</a>. The main steps performed by this tool are:</p>
<ul>
<li>Multiple sequence alignment of our samples against the <em>Wuhan-Hu-1</em> reference genome, using the <code>minimap2</code> software.</li>
<li>Assigning our sequences to lineages based on the current global phylogeny. Two methods/software are available:
<ul>
<li><em>pangoLEARN</em> (default) uses a pre-trained machine learning model.</li>
<li><em>UShER</em> uses a more classic parsimony-based method, but highly optimised for working with large numbers of sequences.</li>
</ul></li>
<li>Classifying our sequences according to the WHO nomenclature of variants of interest/concern using the <code>scorpio</code> software.</li>
</ul>
<p>Because <code>pangolin</code> uses a series of tools and models internally (which change over time as more sequences become available), it is useful to know which versions are being used in the current analysis. To do this, we can run the command:</p>
<pre class="console"><code>$ pangolin --all-versions</code></pre>
<pre><code>pangolin: 3.1.17
pangolearn: 2021-11-25
constellations: v0.0.30
scorpio: 0.3.15
pango-designation used by pangoLEARN/Usher: v1.2.101
pango-designation aliases: 1.2.112</code></pre>
<p>To run <em>Pangolin</em> we can use the following command:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pangolin</span> <span class="at">--outdir</span> directory/of/your/choice path/to/your/sequences.fasta</span></code></pre></div>
<p>This will generate a “comma-separated values” (CSV) file called <code>lineage_report.csv</code> in the output directory that we specify. The file will contain information about the lineage assignment analysis from <em>Pangolin</em>, including WHO variants of concern identified using the <em>Scorpio</em> software. A detailed explanation of the columns of this file is given in the <a href="https://cov-lineages.org/resources/pangolin/output.html">Pangolin documentation page</a>.</p>
<p>One thing to note is that first <em>Pangolin</em> will check if the sequence is identical to a representative sequence of a lineage (a representative sequence is one that is used to train the models that assign sequences to lineages). If it is, then <code>pangoline</code> doesn’t run the lineage assignment step with <em>pangoLearn</em>/<em>Usher</em> nor the variant classification with <em>Scorpio</em>. In these cases, the sequence will have the note “Assigned from designation hash” in the output CSV report file.</p>
<div class="exercise">
<p>Go to the course materials directory <code>03-lineages</code> (on our training machines <code>cd ~/Course_Materials/03-lineages</code>). There you will find a directory called <code>data</code>, which contains a file called <code>all_sequences.fa</code>, containing both UK and India samples that we processed in the <a href="04-artic_nextflow.md">Consensus Assembly section</a>.</p>
<p>To run <code>pangolin</code> on our sequences with default options, we would use the following command:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pangolin</span> <span class="at">--outdir</span> results/pangolin/ data/all_sequences.fa</span></code></pre></div>
<ul>
<li>Create a directory for the output (look at the command above to see what this directory should be called).</li>
<li>Look at <em>Pangolin</em>’s help documentation (<code>pangolin --help</code>) and look for the options you can use to modify the command above to:
<ul>
<li>Use <em>UShER</em> for placing sequences on the phylogeny (instead of the default, which uses <em>pangoLEARN</em>).</li>
<li>Output the multiple sequence alignment generated during the analysis.</li>
<li>Use 8 CPUs (or “threads”) for parallel processing.</li>
</ul></li>
<li>Run the modified command on the terminal. (<strong>Bonus:</strong> save the command in a new shell script for reproducibility.)</li>
<li>Using the file explorer <i class="fa-solid fa-folder"></i>, go to the output directory and open the file named <code>lineage_report.csv</code> (this will open with the spreadsheet software).
<ul>
<li>Check whether there any variants of concern (these are found in the column called <code>scorpio_call</code>).</li>
<li>Check whether there are any problematic samples.</li>
</ul></li>
</ul>
<details>
<summary>
Answer
</summary>
<p>The first thing we should do is create the output directory for our results, and also a directory for our scripts:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/pangolin/</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> scripts/</span></code></pre></div>
<p>We can look at the tool’s help by using <code>pangolin --help</code>. From that documentation we can see the following three options that would do what we want:</p>
<ul>
<li><code>--usher</code> uses the UShER algorithm to place our sequences in the global phylogeny.</li>
<li><code>--alignment</code> outputs the multiple sequence aligned from <code>minimap2</code> as a FASTA file.</li>
<li><code>--threads</code> allows us to use multiple CPUs for the computation.</li>
</ul>
<p>Therefore, the command to run our analysis is:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pangolin</span> <span class="at">--usher</span> <span class="at">--alignment</span> <span class="at">--outdir</span> results/pangolin/ <span class="at">--threads</span> 8 data/all_sequences.fa</span></code></pre></div>
<p>Using a text editor like <code>nano</code>, we could include this command in a script and save it in the <code>scripts</code> folder with the filename <code>pangolin.sh</code> (or another informative name of your choice).</p>
<p>We can open the <code>pangolin_report.csv</code> file on our spreadsheet software to look at the results. From this table, we can make several observations:</p>
<ul>
<li>In the column called “scorpio_call” we can see several samples that were classified as “Alpha” and “Delta” variants.</li>
<li>Looking at the column “status”, we can see that one of the samples failed the quality control applied by <code>pangolin</code>. This is the nanopore sample <code>barcode01</code>, which we had already identified as being slightly problematic, with a high percentage of ambiguous ‘N’ bases.</li>
<li>The samples <code>barcode03</code>, <code>barcode26</code> and <code>barcode27</code> were not assigned a lineage. In this case, the reason is that the <em>Scorpio</em> result disagreed with the lineage assigned by <em>UShER</em>. This could be because of missing data in key identifying mutations that led to a discrepancy between the tools.</li>
<li>Ten of our samples were identical to the representative sequence of certain lineages, which is indicated by the “note” column that says “Assigned from designation hash”. In these cases, <code>pangolin</code> does not run the Scorpio analysis. These are “AY” lineages, which is the Pangolin designation for sub-lineages of B.1.617.2, the Delta variant lineage (see the <a href="https://cov-lineages.org/lineage_list.html">lineage list page</a>). We could later manually annotate these as also being Delta variant.</li>
<li>Some samples have a “conflict” score &gt; 0. In these cases, we can see that UShER found several possible lineages where the sequence could fit. We should take care in checking these results later in our analysis, perhaps complementing them with other tools such as <em>Nextclade</em> or building a phylogeny from our samples.</li>
</ul>
</details>
</div>
<div class="note">
<p><strong>Lineage assignment: <em>USHER</em> or <em>pangoLEARN</em>?</strong></p>
<p><code>pangolin</code> can use two different methods to place our sequences in the background global phylogeny. Both methods are extremely efficient and designed to work with thousands (or millions) of sequences and generally give comparable results. <em>pangoLearn</em> is faster than <em>UShER</em>, but the latter seems to perform better in the presence of missing data [reference needed].</p>
</div>
</div>
<div id="nextclade" class="section level2">
<h2>Nextclade</h2>
<p><img src="https://raw.githubusercontent.com/nextstrain/nextclade/master/docs/assets/nextstrain_logo.svg" alt="Nextstrain" style="float:right;width:10%"></p>
<p>Another system of clade assignment is provided by <code>nextclade</code>, which is part of the broader software ecosystem <a href="https://clades.nextstrain.org/"><em>Nextstrain</em></a>. This is a complementary analysis to what can be obtained with <em>Pangolin</em>, enriching our insights into the data.</p>
<p><em>Nextclade</em> does similar steps as <em>Pangolin</em>, with some differences in the algorithms that are used:</p>
<ul>
<li>Each sequence is aligned with the <em>Wuhan-Hu-1</em> reference genome using a local alignment algorithm.</li>
<li>Samples are placed in the global phylogeny using a distance-based metric (placing the sequence on the tree where it has the highest similarity with).</li>
<li>Clade assignment is done based on the previous phylogeny placement step.</li>
</ul>
<p>You can find more details about <em>Nextclade</em>’s methods on <a href="https://docs.nextstrain.org/projects/nextclade/en/stable/user/algorithm/index.html">its documentation</a>. <em>Nextclade</em> also provides several quality control metrics, which are very useful to identify problematic samples.</p>
<p>There are two main ways to use <em>Nextclade</em>:</p>
<ul>
<li>Through a web interface, which is easier to use, but cannot be configured or automated.</li>
<li>Through a command line interface, which gives us more control over the analysis and can be included in a shell script to automate our analysis.</li>
</ul>
<div id="web-interface" class="section level3">
<h3>Web Interface</h3>
<p><em>Nextclade</em> offers a convenient interface, which can be used to run the default analysis on a set of sequences:</p>
<ul>
<li>Go to <a href="https://clades.nextstrain.org/">nextclade.org</a></li>
<li>Select <strong>SARS-CoV-2</strong> and click <strong>Next</strong></li>
<li>Click <strong>Select a file</strong> to browse your computer and upload the FASTA file with the cleaned consensus sequences (<code>data/all_sequences.fa</code>).</li>
<li>Click <strong>Run</strong></li>
</ul>
<p>Nextclade will show a progress of its analysis at the top of the page, and the results of several quality control metrics in the main panel (see Figure).</p>
<div class="figure">
<img src="images/nextclade_overview.svg" alt="" />
<p class="caption">Overview of the Nextclade web interface.</p>
</div>
<div class="exercise">
<p>Now that we have loaded our consensus sequences to <a href="https://clades.nextstrain.org/">nextclade.org</a>, answer the following questions:</p>
<ol style="list-style-type: decimal">
<li>Are there any samples that were classified as “bad” quality? If so, what is the main reason for their low quality? How does this relate with your previous look at the sequencing coverage in these datasets?
<details>
<summary>
Hint
</summary>
Use the data filters to help you answer this question.
</details></li>
<li>Can you identify any sequences classified as a WHO variant of concern? Does this agree with the previous analysis from <em>Pangolin</em>?
<details>
<summary>
Hint
</summary>
Although this information is available in the main panel, it might be easier to switch to the Phylogeny view, using the button on the top-right.
</details></li>
<li>The image below shows the positions of the mutations that characterise the Delta variant. On the Nextclade browser, select “Protein S” and in the right panel confirm that those mutations occur in the sequences classified as Delta.
<details>
<summary>
Hint
</summary>
To make this exploration easier, try filtering your samples for clade “21”.
</details></li>
</ol>
<div class="figure">
<a href="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/SARS-CoV-2_Delta_variant.svg/6878px-SARS-CoV-2_Delta_variant.svg.png" target="_blank"> <img align="center" src="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/SARS-CoV-2_Delta_variant.svg/6878px-SARS-CoV-2_Delta_variant.svg.png"/> </a>
<p class="caption">
Mutations in the Delta variant Spike protein. The position of each mutation relative to the Wuhan-Hu-1 reference genome is shown. For example: “T19R” means that a T (Threonine) changed to a R (Arginine) in position 19 of the protein; “Δ157-158” means that there is a deletion in position 157-158 of the protein. Click the image to open a larger size. (source: <a href="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/SARS-CoV-2_Delta_variant.svg/6878px-SARS-CoV-2_Delta_variant.svg.png">Wikipedia</a>)
</p>
</div>
<details>
<summary>
Answer
</summary>
<p><strong>A1.</strong></p>
<p>To look at how many samples are classified as different quality categories, we can use the filter button on the top-right <svg stroke="grey" fill="grey" stroke-width="0" viewBox="0 0 512 512" height="15" width="15" xmlns="http://www.w3.org/2000/svg"><path d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"></path></svg> . We can tick/untick based on the way <em>Nextclade</em> classified the sequences in quality categories. By selecting only those samples that have “Bad Quality”, we can see that nearly all of them are from the nanopore data. One of the main errors is due to too much missing data (ambiguous bases, “N”). We can see these as grey boxes in the variant panel on the right.</p>
<p><strong>A2.</strong></p>
<p>In the column <strong>Clade</strong> we can see which clades each sequence was assigned to, following <em>Nextclade</em>’s own nomenclature. However, we also get information about the WHO variants of concern, which use the greek alphabet nomenclature. We can see “Alpha” and “Delta” variants in the data. This is perhaps even better visualised in the phylogeny view, by pressing the <svg width="20" height="25"><g transform="translate(0,2)"><svg width="20" height="20" viewBox="16 19 27 22"><g id="Group" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" transform="translate(17.000000, 20.000000)"><polyline id="Path-3" stroke="#222" points="2.94117647 12.8571429 2.94117647 18.6083984 17.8459616 18.6083984 17.8459616 16.0260882 24.870031 16.0260882"></polyline><polyline id="Path-4" stroke="#222" points="17.8500004 18.5714286 17.8500004 20 25.0119374 20"></polyline><polyline id="Path-5" stroke="#222" points="2.94117647 12.8571429 3.00303284 7.21191058 10.2360102 7.11220045 10.4159559 2.99881623 17.1561911 2.93803403 17.1561911 0 24.313534 0"></polyline><polyline id="Path-6" stroke="#222" points="17.1599998 2.85714286 17.1599998 4.28571429 24.512941 4.28571429"></polyline><path d="M0,12.8571429 L2.94117647,12.8571429" id="Path-6" stroke="#222"></path><polyline id="Path-7" stroke="#222" points="10.2941176 7.14285714 10.2941176 11.4285714 24.5454676 11.4285714"></polyline></g></svg></g></svg> phylogeny button on the top-right.</p>
<p>From there, we can see several variants classifed as “Delta” and “Alpha”. Looking at some specific ones, we can see they are the same samples that <em>Pangolin</em> identified as variants of concern. For example, sample ERR5921254 is classified as “21J (Delta)” by <em>Nextclade</em> and as “Delta (AY.4-like)” by <em>Pangolin/Scorpio</em>.</p>
<p>There were some samples that <em>Pangolin</em> had not assigned a lineage because there was a conflict between <em>UShER</em> and <em>Scorpio</em> (samples <code>barcode03</code>, <code>barcode26</code> and <code>barcode27</code>). <em>Nexclade</em> does classify these sequences as B.1.1.7 (Alpha), which seems to agree with the <em>UShER</em> assignment. However, these samples are highlighted as having too much missing data, which may be the reason why <em>Scorpio</em> did not classify them as a variant of concern.</p>
<p><strong>A3.</strong></p>
<p>Going back to the main panel, we can use the <svg stroke="grey" fill="grey" stroke-width="0" viewBox="0 0 512 512" height="15" width="15" xmlns="http://www.w3.org/2000/svg"><path d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"></path></svg> filter button to show only sequences from clade “21” (this includes “21J”, “21I” and “21A”, which are all sub-clades of <em>Delta</em>).</p>
<p>Comparing the mutation panel on the right with the mutations available from <em>Wikipedia</em>, we can see that they are all present in our sequences. Some other mutations seem to be more specific to the sub-clades. For example “A222V” seems to be specific to clade “21I”.</p>
</details>
</div>
<div class="note">
<p>When using the <em>Nextclade</em> web tools, the data does not leave your computer, so privacy concerns are not an issue.</p>
</div>
</div>
<div id="command-line-interface" class="section level3">
<h3>Command-line Interface</h3>
<p>To use the command line interface of <em>Nextclade</em> we need to first download “background” data, which will be used to classify and integrate our samples in the global phylogeny.</p>
<p>We can check the available data for SARS-CoV-2 by running:</p>
<pre class="console"><code>$ nextclade dataset list --name sars-cov-2</code></pre>
<p>The output of this command will inform us that we can run the following command to obtain the current (most up-to-date) version of the background dataset:</p>
<pre class="console"><code>$ nextclade dataset get --name sars-cov-2 --output-dir resources/nextclade_dataset</code></pre>
<p>This command downloads data used by <code>nextclade</code> to compare our sequences with a pre-computed tree of publicly available sequences (so-called <a href="https://docs.nextstrain.org/projects/nextclade/en/stable/user/terminology.html#reference-tree-concept">reference tree</a>).</p>
<p>Now that we have downloaded our data, we are ready to run our <code>nextclade</code> analysis:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nextclade</span> run <span class="dt">\</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>   <span class="at">--in-order</span> <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>   <span class="at">--input-fasta</span> data/all_sequences.fa <span class="dt">\</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>   <span class="at">--input-dataset</span> resources/nextclade_dataset/ <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>   <span class="at">--include-reference</span> <span class="dt">\</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>   <span class="at">--output-dir</span> results/nextclade <span class="dt">\</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>   <span class="at">--output-basename</span> uk_india <span class="dt">\</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>   <span class="at">--output-json</span> results/nextclade/uk_india.json <span class="dt">\</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>   <span class="at">--output-tsv</span> results/nextclade/uk_india.tsv <span class="dt">\</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>   <span class="at">--output-tree</span> results/nextclade/uk_india.auspice.json <span class="dt">\</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>   <span class="at">--min-length</span> 27000</span></code></pre></div>
<p>The result from this analysis will generate several files in the output folder <code>results/nextclade</code>:</p>
<ul>
<li><code>uk_india.aligned.fasta</code> - aligned nucleotide sequences.</li>
<li><code>uk_india.gene.*.fasta</code> - aligned peptides.</li>
<li><code>uk_india.tsv</code> - analysis results, including clade assignment and several quality-control metrics.
<ul>
<li><code>uk_india.json</code> - analysis results in a machine-readable format.</li>
</ul></li>
<li><code>uk_india.auspice.json</code> - phylogenetic tree in a format that is compatible with <a href="https://auspice.us/"><em>Auspice</em></a> for interactive visualisation.</li>
<li><code>uk_india.insertions.csv</code> - nexclade removes insertions so that they no longer appear in the aligned peptide sequences. This CSV file contains the position and sequence of all those insertions.</li>
<li><code>uk_india.errors.csv</code> - CSV file with of errors and warnings for each sequence.</li>
</ul>
<p>These files are equivalent to what the web-tool we used in the previous section gave us as the results.</p>
<p>We can visualise the result of this analysis by uploading the <code>*.auspice.json</code> file to <a href="https://auspice.us/">auspice.us</a>. In addition, we can upload a metadata table, which can be helpful to further explore our data.</p>
<div class="exercise">
<p>Go to <a href="https://auspice.us/">auspice.us</a> and drag-and-drop the <code>uk_india.auspice.json</code> file onto the browser window. This should open a phylogeny panel, similar to what we previously obtained with the Web interface.</p>
<p>Drag-and-drop our sample metadata file (stored in <code>sample_info_nextclade.csv</code>) into the browser window. This will add this metadata to the interactive view, including a new map panel showing the location of our samples.</p>
<p>Take some time to explore your data using this tool.</p>
</div>
<div class="note">
<p>Nextclade adds the suffix “_new” to our sample names, so we created a new metadata file where we changed the name of our samples to include “_new” in their name as well. This is needed when adding metadata to our results on <a href="https://auspice.us/">auspice.us</a>. Apart from that, the file is exactly the same as the <code>sample_info.csv</code> file we have been using so far.</p>
</div>
<div class="exercise">
<p><em>Nextclade</em> also outputs the protein sequences for each of the SARS-CoV-2 genes. We can visualise these with the program AliView.</p>
<p><img src="https://avatars.githubusercontent.com/u/7050126?s=200&v=4" alt="Nextstrain" style="float:right;width:5%"></p>
<ul>
<li>Open AliView (this program’s icon is shown on the right)</li>
<li>Go to “File &gt; Open File” and navigate to “Course_Materials &gt; 03-lineages &gt; results &gt; nextclade” to open the file “uk_india.gene.S.fasta”</li>
<li>Can you see the mutations characteristic of the Delta variant? (see image below)</li>
</ul>
<div class="figure">
<a href="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/SARS-CoV-2_Delta_variant.svg/6878px-SARS-CoV-2_Delta_variant.svg.png" target="_blank"> <img align="center" src="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/SARS-CoV-2_Delta_variant.svg/6878px-SARS-CoV-2_Delta_variant.svg.png"/> </a>
<p class="caption">
Mutations in the Delta variant Spike protein. Click image to open larger size. (source: <a href="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/SARS-CoV-2_Delta_variant.svg/6878px-SARS-CoV-2_Delta_variant.svg.png">Wikipedia</a>)
</p>
</div>
<details>
<summary>
Answer
</summary>
<p>AliView allows us to interactively visualise a FASTA file with protein or DNA alignments.</p>
<p>To help us explore these data, we can right-click the first sequence (Wuhan-Hu-1 reference) and choose “Set this sequence as template when Highlighting difference”. This will make it easier to see aminoacid differences in our samples.</p>
<p>We can scroll to the right to see that several samples contain the mutation at position 19 changing a T to an R, at position 157 there is a deletion and a change to a G, at position 452 a change from L to R, etc.</p>
<p>If we look at these samples’ names, they match the classification obtained with Nextclade and Pangolin as being the Delta variant.</p>
</details>
</div>
<!--
The alignment between _Nextclade_ and _Pangolin_ is slightly different at the start and end of the sequences. 
This is because _Pangolin_ masks the first 265bp and last 229bp of each sequence (confirm - empirically this seems to be the case; presumably this is done because assembly quality is generally lower there?).
-->
</div>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<div class="highlight">
<p><strong>Key Points</strong></p>
<ul>
<li>Groups of similar SARS-CoV-2 sequences are classified into lineages or clades by different groups. The main nomenclature systems in use are <em>Nextstrain</em>, <em>Pangolin</em> and <em>GISAID</em>.</li>
<li>In addition, the World Health Organisation (WHO) classifies some forms of SARS-CoV-2 as <em>variants of concern</em> or <em>variants of interest</em>. These are forms of the virus that have been determined to have a significant public health impact.</li>
<li>The software <code>pangolin</code> can be used to assign consensus sequences to lineages and WHO variants.</li>
<li><em>Nextclade</em> can also be used to assign sequences to clades and WHO variant classification. <em>Nextclade</em> can be run from a web interface and the command line (<code>nextclade</code>).</li>
</ul>
</div>
</div>
</div>

<!DOCTYPE html>
<html>
<head>
<style>
* {
  box-sizing: border-box;
}

.img-container-left {
  float: left;
  width: 33.33%;
  padding: 15px;
}

.img-container-right {
  float: right;
  width: 33.33%;
  padding: 15px;
}

.clearfix::after {
  content: "";
  clear: both;
  display: table;
}
</style>
</head>
<body>

<hr>
<div class="clearfix">
  <div class="img-container-left">
    <img src="assets/img/logo_btf.png" alt="UoC Bioinformatics Training Facility" style="width:100%">
  </div>
  <div class="img-container-right">
    <img src="assets/img/logo_ukhsa.svg" alt="UK Public Health England" style="width:40%">
  </div>
</div>

</body>
</html>





</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
