<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>SARS-CoV-2 Genomics</title>

<script src="site_libs/header-attrs-2.16/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="assets/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">UoC BioinfoTraining</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="101-setup.html">
    <span class="fa fa-gear"></span>
     
    Setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-file-lines"></span>
     
    Materials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">The Unix Shell</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="02a-unix_intro.html">Intro to Unix</a>
        </li>
        <li>
          <a href="02b-unix_files_directories.html">Files and Directories</a>
        </li>
        <li>
          <a href="02c-unix_text_manipulation.html">Text Manipulation</a>
        </li>
        <li>
          <a href="02d-unix_pipes.html">Combining Commands &amp; Writing Scripts</a>
        </li>
      </ul>
    </li>
    <li>
      <a href="01-intro.html">SARS-CoV-2 Genomic Surveillance</a>
    </li>
    <li>
      <a href="03-intro_ngs.html">Intro to NGS Data</a>
    </li>
    <li>
      <a href="04-consensus.html">Consensus Assembly</a>
    </li>
    <li>
      <a href="05-lineage_analysis.html">Lineages and Variants</a>
    </li>
    <li>
      <a href="06-phylogeny.html">Building Phylogenies</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Extras
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="102-unix_cheatsheet.html">Unix Cheatsheet</a>
    </li>
    <li>
      <a href="106-file_formats.html">Common File Formats</a>
    </li>
    <li>
      <a href="104-wsl_windows.html">Windows Subsystem for Linux</a>
    </li>
    <li>
      <a href="105-vs_code.html">Visual Studio Code</a>
    </li>
    <li>
      <a href="103-tools_and_resources.html">SARS-CoV-2 Resources</a>
    </li>
  </ul>
</li>
<li>
  <a href="100-homework.html">
    <span class="fa fa-pencil"></span>
     
    Homework
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cambiotraining/sars-cov-2-genomics">
    <span class="fa fa-brands fa-github"></span>
     
    Github
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<div id="building-a-phylogenetic-tree" class="section level1">
<h1>Building a Phylogenetic Tree</h1>
<div class="highlight">
<p><strong>Questions</strong></p>
<ul>
<li>How can I produce multiple sequence alignment of the consensus sequences?</li>
<li>How can I build a phylogenetic tree from my consensus sequences?</li>
<li>How can I scale my tree based on sampling dates?</li>
</ul>
<p><strong>Learning Objectives</strong></p>
<ul>
<li>Understand the basics of how phylogeny trees are constructed using maximum likelihood methods.</li>
<li>Describe what branch lengths represent in typical SARS-CoV-2 phylogenetic tree.</li>
<li>Recognise the limitations and challenges in building phylogenies for SARS-CoV-2.</li>
<li>Use <em>MAFFT</em> to produce a multiple sequence alignment.</li>
<li>Use <em>IQ-Tree</em> for phylogenetic tree inference.</li>
<li>Use <em>TimeTree</em> to obtain time-scaled phylogenetic trees using sample collection date metadata.</li>
</ul>
</div>
<div class="note">
<p>This section has an accompanying <a href="https://drive.google.com/file/d/19MCiM8JhWgtSwvQdotfJRlyv7Z7CVJwz/view?usp=sharing" target="_blank">slide deck</a>.</p>
</div>
<!--
## Phylogenetics Basics

TODO: 

- brief intro to phylogenetics, based on Nicola's presentation. 
- Mention interpretation of branch length in particular SARS-CoV-2 has ~2 mutations per month (probably replication errors), leading to very short branch lengths and polytomies. 
- Note about bootstraping - should probably be avoided, since there are usually too few sites for meaningful results. 


- Multifurcations happen when several samples have the same sequence, so the branch lengths between them are zero, so they all collapse to a single node. This can happen for example if you have a local outbreak of a variant where every sample has the same variant.
- Recombination can cause problems in phylogenies. This is when two viral sequences recombine with each other, so now the new sequence has part of its genome closely-related to one parent and the other from the other parent. This can cause problems with phylogenetic trees. There are software that can detect recombination events. In particular the software RIPPLES can work with large sample sizes.

### Alignment

To build a phylogenetic tree we need to start with all our sequences aligned to each other. 
The alignment might result in _gaps_ in some sequences, where missing nucleotides in some sequences are represented with `-` character. 
The alignment is usually stored as a FASTA file, and all sequences have to be of the same length (because they are aligned to each other). 

Gaps (`-`) and ambiguous positions (`N`) are treated as missing data by the phylogenetic methods. 

Something to note is that phylogenetic inference assumes that the alignments we give are correct. 
However, errors in the consensus sequence or missing data can sometimes lead to poor alignments.
One thing to pay attention to is whether there are very long branches in our trees, suggesting a sample with an unusual high number of mutations. 
Inspecting these samples may reveal errors in the alignment, for example due to high proportion of missing data that leads to poor alignments, which may result in false variable positions. 

Other errors are harder to detect but work from Turakaia et al was able to build a collection of recurrent errors that affect phylogenies.


### Tree Inference

There are broadly three types of methods: 

- distance
- parsimony
- maximum likelihood

#### Distance Phylogenies

These methods rely on calculating the distance between each pair of sequences.
Popular methods of this kind are _neighbour-joining_ (NJ) or 

#### Maximum Parsimony

Infers a tree that requires the fewest number of mutations needed along each branch of the tree to explain the data. 

This method can be inaccurate when there are long branches in the tree (this problem is referred to as "long branch attaction"). 
However, for SARS-CoV-2 this is not a big problem, because the sequences tend to be very similar to each other. 

In fact, a recent tool called _UShER_ has been developed for SARS-CoV-2 analysis using a parsimony-based method. 
This tool was developed to be very efficient at working with millions of samples, which is harder with maximum likelihood methods. 

#### Maximum Likelihood

- Uses a probabilistic model for genome evolution.
- Model of evolution:
  - DNA substituion model
    - JC69 assumes only one mutation rate
    - HKY85 assumes different mutation rates (transitions have different rates)
    - GTR is another one
  - Rate heterogeneity:
    - Invariant sites - the model will assume that a certain proportion of the sites in the genome might never change.
    - Rate variation (Gamma models used) - the model will assume that different sites of the genome might evolve at different rates.
- Generally, more complex models give better results (but are more computationally demanding and require more data)
-  
-->
<div id="sars-cov-2-phylogeny" class="section level2">
<h2>SARS-CoV-2 Phylogeny</h2>
<p><img src="https://raw.githubusercontent.com/roblanf/sarscov2phylo/master/tree_image.jpg" alt="Global Phylogeny" style="float:right;width:20%"></p>
<p>Building phylogenetic trees for SARS-CoV-2 is challenging due to the large number of sequences available, with <a href="https://www.gisaid.org/index.php?id=208">millions of genomes submited to GISAID</a> to this day. This means that using maximum-likelihood inference tools to build a global SARS-CoV-2 phylogeny is both time-consuming and computationally demanding.</p>
<p>However, several researchers have dedicated their time to identifying tools and models suitable for the phylogenetic analysis of this organism. For example, the <a href="https://github.com/roblanf/sarscov2phylo">global phylogenies repository</a> from Rob Lanfear provide with several tips on building phylogenetic trees for this organism. Their trees are regularly updated and available to download from the GISAID website (which requires an account).</p>
<p>Global phylogenies are also available from the groups of Russell Corbett-Detig and Yatish Turakhia, who have developed <a href="https://doi.org/10.1093/molbev/msab264">efficient methods and tools</a> for dealing with large phylogenies. These tools include <em>UShER</em> and <em>matUtils</em>, introducing a new and efficient file format for storing phylogenetic trees, mutations and other annotations (such as lineages) called <em>mutation-annotated trees</em> (MAT format). Their phylogenies are updated daily and are <a href="http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/">publicly available for download</a>. (Note: Course materials covering these tools are still under development.)</p>
<p>Two popular tools used for phylogenetic inference via maximum-likelihood are <strong><em>FastTree</em></strong> and <strong><em>IQ-Tree</em></strong>. Generally, when building a tree from a collection of samples you can include the Wuhan-Hu-1 reference sequence as an outgroup to root the tree. Optionally, you can also add a collection of sequences from around the world (and across time) to contextualise your samples in the global diversity.</p>
<p>As an input, these programs need a <em>multiple sequence alignment</em> FASTA file, which is where we start our analysis.</p>
<div class="note">
<p><strong>Data for this section</strong></p>
<p>We will work from the course materials folder called <code>04-phylogeny</code>, which contains the following files:</p>
<ul>
<li><code>data/uk_consensus.fa</code> and <code>data/india_consensus.fa</code> are consensus sequences from the UK and India, respectively. These are the sequences previously assembled using the <code>nf-core/viralrecon</code> pipeline.</li>
<li><code>sample_annotation.tsv</code> is a tab-separated values (TSV) file with information about each sample such as the date of collection and lineage/clade they were assiged to from our analysis. We will use this table to annotate our phylogenetic trees. This table can also be opened in a spreadsheet program such as Excel.</li>
<li><code>resources/reference/sarscov2.fa</code> is the reference genome.</li>
</ul>
</div>
</div>
<div id="alignment" class="section level2">
<h2>Alignment</h2>
<p>The first step in building a phylogenetic tree is to produce a multiple sequence alignment from all our consensus sequences. This is the basis for building a phylogenetic tree from the positions that are variable across samples.</p>
<p>A widely used multiple sequence alignment software is called <strong><em>MAFFT</em></strong>. For SARS-CoV-2, a reference-based alignment approach is often used, which is suitable for closely-related genomes. MAFF provides this functionality, which is detailed in its <a href="https://mafft.cbrc.jp/alignment/software/closelyrelatedviralgenomes.html">documentation</a>.</p>
<p>We demonstrate the analysis using the UK samples in our course materials. First, start by creating a directory for the output:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/mafft</span></code></pre></div>
<p>Then, we run the command to generate a reference-based alignment:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mafft</span> <span class="at">--6merpair</span> <span class="at">--maxambiguous</span> 0.2 <span class="at">--addfragments</span> data/uk_consensus.fa resources/reference/sarscov2.fa <span class="op">&gt;</span> results/mafft/uk_alignment.fa</span></code></pre></div>
<p>The meaning of the options used is:</p>
<ul>
<li><code>--6merpair</code> is a fast method for estimating the distances between sequences, based on the number of short 6bp sequences shared between each pair of sequences. This is less accurate than other options available (like <code>--localpair</code> and <code>--globalpair</code>), but runs much faster in whole genome data like we have.</li>
<li><code>--maxambiguous 0.2</code> automatically removes samples with more than 20% ambiguous ‘N’ bases (or any other value of our choice). This is a convenient way to remove samples with poor genome coverage from our analysis.</li>
<li><code>--addfragments data/consensus_sequences.fa</code> is the FASTA file with the sequences we want to align.</li>
<li>finally, at the end of the command we give the reference genome as the input sequence to align our sequences against.</li>
</ul>
<p>MAFFT provides several other methods for alignment, with tradeoffs between speed and accuracy. You can look at the full documentation using <code>mafft --man</code> (<code>mafft --help</code> will give a shorter description of the main options). For example, from its documentation we can see that the most precise alignment can be obtained with the options <code>--localpair --maxiterate 1000</code>. However, this is quite slow and may not be feasible for whole genome data of SARS-CoV-2.</p>
<div id="visualising-alignments" class="section level3">
<h3>Visualising alignments</h3>
<p>We can visualise our alignment using the software <a href="https://ormbunkar.se/aliview/">AliView</a>, which is both lightweight and fast, making it ideal for large alignments. Visualising the alignment can be useful for example to identify regions with missing data (more about this below).</p>
<div class="figure">
<img src="images/alignment_aliview.png" alt="" />
<p class="caption">Snapshop of an alignment visualised with AliView. In this case we are looking at the end of the alignment of our sequences, which shows a typical high number of missing (‘N’) bases.</p>
</div>
<div class="note">
<p><strong>Other Alignment Strategies</strong></p>
<p>There are other commonly used alignment tools used for SARS-CoV-2 genomes:</p>
<ul>
<li>The <a href="https://lh3.github.io/minimap2/"><code>minimap2</code></a> software has been designed for aligning long sequences to a reference genome. It can therefore be used to align each consensus sequence to the Wuhan-Hu-1 genome. This is the tool internally used by <em>Pangolin</em>.</li>
<li><em>Nextclade</em> uses an internal alignment algorithm where each consensus sequence is aligned with the reference genome. The alignment produced from this tool can also be used for phylogenetic inference.</li>
</ul>
<p>It is worth mentioning that when doing reference-based alignment, insertions relative to the reference genome are not considered.</p>
</div>
</div>
</div>
<div id="tree-inference-iq-tree" class="section level2">
<h2>Tree Inference: IQ-Tree</h2>
<p><em>IQ-TREE</em> supports <a href="http://www.iqtree.org/doc/Substitution-Models">many substitution models</a>, including models with <em>rate heterogeneity</em> across sites.</p>
<p>Let’s start by creating an output directory for our results:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/iqtree</span></code></pre></div>
<p>And then run the program with default options (we set <code>--prefix</code> to ensure output files go to the directory we just created and are named “uk”):</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> results/mafft/uk_alignment.fa <span class="at">--prefix</span> results/iqtree/uk</span></code></pre></div>
<p>Without specifying any options, <code>iqtree2</code> uses <a href="https://www.nature.com/articles/nmeth.4285"><em>ModelFinder</em></a> to find the substituion model that maximizes the likelihood of the data, while at the same time taking into account the complexity of each model (using information criteria metrics commonly used to assess statistical models).</p>
<p>From the information printed on the console after running the command, we can see that the chosen model for our alignment was “GTR+F+I”, a <a href="https://en.wikipedia.org/wiki/Substitution_model#Generalised_time_reversible"><em>generalised time reversible</em> (GTR) substitution model</a>. This model requires an estimate of each base frequency in the population of samples, which in this case is estimated by simply counting the frequencies of each base from the alignment (this is indicated by “+F” in the model name). Finally, the model includes rate heterogeneity across sites, allowing for a proportion of invariant sites (indicated by “+I” in the model name). This makes sense, since we know that there are a lot of positions in the genome where there is no variation in our samples.</p>
<p>We can look at the output folder (specified with <code>--prefix</code>) where we see several files with the following extension:</p>
<ul>
<li><code>.iqtree</code> - a text file containing a report of the IQ-Tree run, including a representation of the tree in text format.</li>
<li><code>.treefile</code> - the estimated tree in NEWICK format. We can use this file with other programs, such as <em>FigTree</em>, to visualise our tree.</li>
<li><code>.log</code> - the log file containing the messages that were also printed on the screen.</li>
<li><code>.bionj</code> - the initial tree estimated by neighbour joining (NEWICK format).</li>
<li><code>.mldist</code> - the maximum likelihood distances between every pair of sequences.</li>
<li><code>ckp.gz</code> - this is a “checkpoint” file, which IQ-Tree uses to resume a run in case it was interrupted (e.g. if you are estimating very large trees and your job fails half-way through).</li>
<li><code>.model.gz</code> - this is also a “checkpoint” file for the model testing step.</li>
</ul>
<p>The main files of interest are the report file (<code>.iqtree</code>) and the NEWICK tree file (<code>.treefile</code>).</p>
<div class="note">
<p><strong>Inference of very large trees</strong></p>
<p>Although running <em>IQ-Tree</em> with default options is fine for most applications, there will be some bottlenecks once the number of samples becomes too large. In particular, the <em>ModelFinder</em> step may be very slow and so it’s best to set a model of our choice based on other people’s work. For example, <a href="https://github.com/roblanf/sarscov2phylo/blob/13-11-20/tree_estimation.md#which-model-is-best">work by Rob Lanfear</a> suggests that models such as “GTR+G” and “GTR+I” are suitable for SARS-CoV-2. We can specify the model used by <code>iqtree2</code> by adding the option <code>-m GTR+G</code>, for example.</p>
<p>For very large trees (over 10k or 100k samples), using an alternative method to place samples in an existing phylogeny may be more adequate. <a href="https://usher-wiki.readthedocs.io/en/latest/"><em>UShER</em></a> is a popular tool that can be used to this end. It uses a parsimony-based method, which tends to perform well for SARS-CoV-2 phylogenies.</p>
</div>
<div id="visualising-trees" class="section level3">
<h3>Visualising Trees</h3>
<p>There are many programs that can be used to visualise phylogenetic trees. In this course we will use <em>FigTree</em>, which has a simple graphical user interface.</p>
<p>To open the tree, go to <kbd><kbd>File</kbd> &gt; <kbd>Open…</kbd></kbd> and browse to the folder with the IQ-Tree output files. Select the file with <code>.treefile</code> extension and click <kbd>Open</kbd>. You will be presented with a visual representation of the tree.</p>
<p>We can also import a “tab-separated values” (TSV) file with annotations to add to the tree. For example, we can use our results from <em>Pangolin</em> and <em>Nextclade</em>, as well as other metadata to improve our visualisation (we have prepared a TSV with this information combined, which you could do using Excel or another spreadsheet software).</p>
<ul>
<li>Go to <kbd><kbd>File</kbd> &gt; <kbd>Import annotations…</kbd></kbd> and open the annotation file.</li>
<li>On the menu on the left, click <kbd>Tip Labels</kbd> and under “Display” choose one of the fields of our metadata table. For example, you can display the lineage assigned by <em>Pangolin</em> (“pango_lineage” column of our annotation table).</li>
</ul>
<p>There are many ways to further configure the tree, including highlighting clades in the tree, and change the labels. See the figure below for an example.</p>
<div class="figure">
<img src="images/phylogeny_figtree.svg" alt="" />
<p class="caption">Annotated phylogenetic tree obtained with FigTree. We used the lineage of each sample as our tip labels and aligned the labels on the right (check the tickbox at the top of the left menu called “Align tip labels”). We identified two clades in the tree that corresponded to the Alpha and Delta variants, and used the “Highlight” tool to give them different colours. To do this, change the “Selection Mode” at the top to “Clade”, then select the branch at the base of the clade you want to highlight, and press the “Highlight” button on the top to pick a colour.</p>
</div>
<!-- 
From [here](https://github.com/roblanf/sarscov2phylo/issues/20#issuecomment-855319059):

```bash
iqtree2 -s aln.fasta -t NJ-R -n 0 -m GTR+R4 -nt 8 -blmin 0.00000000001
```

Although if we run with default options, it will use _ModelFinder Plus_ to identify best model (based on AIC):

```bash
iqtree -s aln.fasta -nt 8 -blmin 0.00000000001
```

This takes quite a while to run though. 

NOTE:
- In `iqtree2` the option `-nt` (number threads) is now `-T`. The default auto-detects so we could leave it out.
- Running `fasttree` results in a strange tree with a "laddering" effect. Running `iqtree` with `-blmin` set to a small number solves this somewhat.


### FastTree

From [here](https://github.com/roblanf/sarscov2phylo/commit/55cbb8ddaddaceb2ecd0577f105a3fd37f13d304), I think these are run with performance in mind:

```bash
fasttree -nt -gamma -nosupport \
  -sprlength 500 -nni 0 -spr 5 \
  -refresh 0.8 -topm 1.5 -close 0.75 -noml \
  "$INPUT_FASTA" > "$INPUT_FASTA"'1.multi.fasttree'
    
fasttree -nt -gamma -sprlength 200 -spr 5 -intree \
  "$INPUT_FASTA"'1.multi.fasttree' \
  "$INPUT_FASTA" > "$INPUT_FASTA"'multi.fasttree'
```

In this document there's a much simpler command, which is probably fine for the purpose of our course: 

```bash
fasttree -nt -nosupport input.fa > output.tree 2> output.log
```

**NOTE to self:** ran this with additional `-gtr` option and took 6 minutes. 

- `-nt` for nucleotide alignment (default is protein alignment).
- `-nosupport` to turn off support values (bootstrap) - it makes sense to turn this off, since we often have very few variable sites to bootstrap from.

We can visualise the tree we just produced with _FigTree_. 
We can also import the output from _Pangolin_ to annotate our tree by variant (**note:** unfortunately FigTree wants a TSV file, but pangolin exports CSV. I did this, but maybe there's a less confusing way to do this? `cat results/pangolin/run1_report.csv | sed 's/,/\t/g' > results/pangolin/run1_report.tsv`)


:::note
<details><summary>Installing `fasttree` and `iqtree`</summary>

You can install both these programs using _Conda_ (more about this package manager in [Reproducible Workflows](08-nextflow_conda_setup.html)).
For example, we can create a new environment called "phylogenetics" with both these programs as well as _FigTree_ (for tree visualisation).

```console
$ mamba create --name phylogenetics fasttree iqtree figtree
```

We can then activate this environment by running `conda activate phylogenetics`, making the programs available to use. 

Note that _FigTree_ is also available on Windows and Mac (see the [software page](http://tree.bio.ed.ac.uk/software/figtree/) for download links).

</details>
:::


:::exercise

- Looking at the tree we just constructed, do you think the Delta variant emerged from the Alpha variant?
- Produce a new phylogenetic tree but this time using both the sequences we processed and a sample of other public sequences that represent the global diversity of SARS-CoV-2 genomes. 
  - First combine the two files into one (`cat`).
  - Produce a combined annotation file with sequence name and lineage.
  - Then run `fasttree` with the same settings we used before.
- Visualise this new tree. Does your conclusion change?

:::

Alignment: MAFFT is used [here](https://github.com/roblanf/sarscov2phylo/blob/master/scripts/global_profile_alignment.sh) `mafft --thread 1 --quiet --keeplength --add $seqfile "$REFERENCE_ALN" > $alfile` 

-->
</div>
</div>
<div id="time-scaled-phylogenies" class="section level2">
<h2>Time-scaled Phylogenies</h2>
<p>The trees that we build from sequence data are scaled using the mutation rate estimated from the sequence alignments. This is useful if we want to know, for example, on average how many mutations separate different branches of the tree.</p>
<p>Another way to scale trees is to use time. For viral genome sequences, we usually have information about their date of collection, and this can be used to scale the phylogeny using the date information. The idea is to rescale the trees such that the x-axis of the tree represents a date rather than number of mutations.</p>
<p>Two programs that can be used to time-scale trees using date information are called <a href="https://treetime.readthedocs.io/en/latest/index.html"><em>TreeTime</em></a> and <a href="https://www.biorxiv.org/content/10.1101/2021.10.27.465994v1"><em>Chronumental</em></a>. <em>Chronumental</em> was developed for dealing with very large phylogenies (millions of samples), but lacks some of the functionalities provided by <em>TreeTime</em> (such as estimating uncertainty in date estimates, reconstructing past sequences, re-rooting trees, among others). So, if you are working with less than ~50,000 sequences, we recommend using <em>TreeTime</em>, otherwise <em>Chronumental</em> is a suitable alternative.</p>
<p>Because we are dealing with a small number of samples, we will show an example of using <em>TreeTime</em> to scale our tree based on our dates:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">treetime</span> <span class="at">--tree</span> results/iqtree/uk.treefile <span class="at">--dates</span> sample_annotation.tsv <span class="at">--aln</span> results/mafft/uk_alignment.fa <span class="at">--outdir</span> results/treetime/uk</span></code></pre></div>
<p>After running, this tool produces several output files in the specified folder. The main files of interest are:</p>
<ul>
<li><code>timetree.nexus</code> is the new date-scaled tree. <a href="https://en.wikipedia.org/wiki/Nexus_file">NEXUS</a> is another tree file format, which can store more information about the tree compared to the simpler NEWICK format. This format is also supported by <em>FigTree</em>.</li>
<li><code>timetree.pdf</code> is a PDF file with the inferred tree, including a time scale on the x-axis. This can be useful for quick visualisation of the tree.</li>
</ul>
<p>We can visualise this tree in <em>FigTree</em>, by opening the file <code>timetree.nexus</code>. The scale of this new tree now corresponds to years (instead of nucleotide substitution rates). We make make several adjustments to this tree, to make it look how we prefer. For example:</p>
<ul>
<li>Label the nodes of the tree with the inferred dates: from the left menu click <kbd>Node Labels</kbd> and under “Display” select “date”.</li>
<li>Import the metadata table (<code>sample_annotation.tsv</code> file) and display the results of <em>Nextclade</em> or <em>Pangolin</em> clades/lineages.</li>
<li>Adjust the scale of the tree to be more intuitive. For example, instead of having the unit of the scale in years, we can change it to months. On the left menu, click <kbd>Time Scale</kbd> and change “Scale factor” to 12 (twelve months in a year). Then click <kbd>Scale Bar</kbd> and change the “Scale range” to 1. The scale now represents 1 month of time, which may be easier to interpret in this case.</li>
</ul>
<p>Note that there is uncertainty in the estimates of the internal node dates from <em>TimeTree</em> and these should be interpreted with some caution. The inference of the internal nodes will be better the more samples we have, and across a wider range of times. In our specific case we had samples all from a similar period of time, which makes our dating of internal nodes a little poorer than might be desired.</p>
<p>More precise tree dating may be achieved by using public sequences across a range of times or by collecting more samples over time. Time-scaled trees of this sort can therefore be useful to infer if there is a recent spread of a new clade in the population.</p>
<div class="exercise">
<p>So far we have focused our analysis on the samples from the UK. In this exercise you will be able to practice these steps on the samples from India. The steps are similar to what we have done so far, and you can consult the materials in the sections above to go through each exercise.</p>
<p>In the following exercises, you can run the commands directly from the command line. But if you feel comfortable with <code>nano</code>, as a bonus, you can try to save the commands in a <em>shell script</em>.</p>
<ol style="list-style-type: decimal">
<li>Using <code>mafft</code>, produce a multiple-sequence alignment from the India consensus sequences in the file <code>data/india_consensus.fa</code>. Save the output in a file named <code>results/mafft/india_alignment.fa</code>.</li>
<li>Using <code>iqtree2</code>, infer a phylogenetic tree from this alignment. Save the output with prefix <code>results/iqtree/india</code>.
<ul>
<li>What substitution model did <em>IQ-Tree</em> infer as the most likely for the data?</li>
</ul></li>
<li>Using <em>FigTree</em>, visualise the inferred tree:
<ul>
<li>Open the <code>.tree</code> output file.</li>
<li>Import the annotation table <code>sample_annotation.tsv</code>.</li>
<li>Make the tip labels display the Nextclade clade instead of the sample names.</li>
<li>Highlight any clusters of the tree containing WHO variants of concern.</li>
</ul></li>
<li>Using <code>timetree</code>, re-scale the phylogeny using dates (the file <code>sample_annotation.tsv</code> can be used as input to <code>timetree</code> along with the previously-created phylogeny and alignments). Output the result to <code>results/treetime/india</code>
<ul>
<li>Open the <code>.nexus</code> output file in <em>FigTree</em>.</li>
<li>Make the node labels display the date inferred by <code>timetree</code>.</li>
</ul></li>
<li>Two samples in the time-scaled tree appear at the root of the tree: IN05 and IN33. But we would have expected the reference sample (MN908947.3) to be the root of the tree, as it was collected in Dec 2019.
<ul>
<li>Investigate what lineages these samples were assigned to.</li>
<li>Go to <a href="https://cov-lineages.org/lineage_list.html" class="uri">https://cov-lineages.org/lineage_list.html</a> and check when those lineages were detected.</li>
<li>Is the collection date metadata for these samples compatible with the lineage information from the <em>Pangolin</em> website? Can you hypothesise what may have happened with these samples?</li>
</ul></li>
</ol>
<details>
<summary>
Answer
</summary>
<p><strong>Question 1</strong></p>
<p>We run our alignment using the following command:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mafft</span> <span class="at">--6merpair</span> <span class="at">--maxambiguous</span> 0.2 <span class="at">--addfragments</span> data/india_consensus.fa resources/reference/sarscov2.fa <span class="op">&gt;</span> results/mafft/india_alignment.fa</span></code></pre></div>
<p>We could optionally visualise our alignment using <em>AliView</em> to check that the alignment was successful.</p>
<p><strong>Question 2</strong></p>
<p>We can fit a maximum-likelihood tree using the following command:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">iqtree2</span> <span class="at">-s</span> results/mafft/india_alignment.fa <span class="at">--prefix</span> results/iqtree/india</span></code></pre></div>
<p>This command prints a lot of output on the screen. We can see all this information (and more!) on the output file <code>results/iqtree/india.iqtree</code>, which contains a log of all the analysis done by the program. We could, for example, use the <code>less</code> program to look inside this file. When we do this, we can see the following:</p>
<pre><code>ModelFinder
-----------

Best-fit model according to BIC: GTR+F+I</code></pre>
<p>Which indicates that the substitution model used by <em>IQ-Tree</em> was “GTR+F+I”. This is the same model that was determined for the UK samples, and an explanation of this model is given in the materials above.</p>
<p><strong>Question 3</strong></p>
<p>Using <em>FigTree</em>:</p>
<ul>
<li>Go to <kbd><kbd>File</kbd> &gt; <kbd>Open…</kbd></kbd> and browse to the folder with the <em>IQ-Tree</em> output files.</li>
<li>Select the file with <code>india.treefile</code> and click <kbd>Open</kbd>.</li>
<li>Go to <kbd><kbd>File</kbd> &gt; <kbd>Import annotations…</kbd></kbd> and open the annotation file <code>sample_annotation.tsv</code>.</li>
<li>On the menu on the left, click <kbd>Tip Labels</kbd> and under “Display” choose the field “nextclade_clade”.</li>
</ul>
<p>On this tree, there is a small group of samples classified as “20I (Alpha; V1)”. These samples correspond to the <em>Alpha</em> variant of concern. To highlight these:</p>
<ul>
<li>Change the “Selection Mode” at the top to “Clade”.</li>
<li>Select the branch corresponding to the base of the group of samples classified as <em>Alpha</em>. This should highlight all those branches.</li>
<li>Click the “Highlight” button at the top and choose a colour.</li>
</ul>
<p>The final result should look similar to what is shown here.</p>
<div class="figure">
<img src="" alt="" />
<p class="caption">TODO</p>
</div>
<p><strong>Question 4</strong></p>
<p>The command to time-scale the tree is:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">treetime</span> <span class="at">--tree</span> results/iqtree/india.treefile <span class="at">--dates</span> sample_annotation.tsv <span class="at">--aln</span> results/mafft/india_alignment.fa <span class="at">--outdir</span> results/treetime/india</span></code></pre></div>
<p>Once complete, we can open the <code>india.nexus</code> tree with <em>FigTree</em>. We can annotate the internal nodes of the tree with the dates inferred by <code>treetime</code> by clicking on the <kbd>Node Labels</kbd> menu on the left and selecting “Display” to be “date”. This should result in a tree similar to the one shown here.</p>
<div class="figure">
<img src="" alt="" />
<p class="caption">TODO</p>
</div>
<p><strong>Question 5</strong></p>
<p>To investigate this strange result, we open the <code>sample_annotation.tsv</code> file in our spreadsheet program. We can see the following information for these two samples:</p>
<pre><code>name    date           country  year     month  pango_lineage      nextclade_clade
IN05    2021-01-21   India    2021   1      A.23.1           19B
IN33    2021-01-21   India    2021   1      A.23.1           19B</code></pre>
<p>Both these samples were assigned to <em>Pango</em> lineage “A.23.1”. From the <a href="https://cov-lineages.org/lineage_list.html">cov-lineages.org</a> website we can see that these were detected as early as 2020-06-08.</p>
<p>However, our samples’ metadata indicate that these samples were collected in January 2021, that is 7 months after these samples were first detected globally.</p>
<p>This contractiction explains the strange result in our time-scaled tree. These samples will have mutations that are part of the older lineage “A”, but were annotated to be from 2021. So <code>treetime</code> put them down at the root of the tree.</p>
<p>This discrepancy could indicate an issue with the metadata collection, which perhaps is wrong. For example, it could be that these samples were collected months before January 2021, but only sequenced then. And by mistake the date of sequencing was recorded as the date of collection.</p>
<p>This is an example how important accurate metadata is for our analysis.</p>
</details>
</div>
</div>
<div id="missing-data-problematic-sites" class="section level2">
<h2>Missing Data &amp; Problematic Sites</h2>
<p>So far, we have been using all of our assembled samples in the phylogenetic analysis. However, we know that some of these have poorer quality (for example the “IN01” sample from India had low genome coverage). Although, generally speaking, sequences with missing data are unlikely to substantially affect the phylogenetic results, their placement in the phylogeny will be more uncertain (since several variable sites may be missing data). Therefore, for phylogenetic analysis, it is best if we remove samples with low sequencing coverage, and instead focus on high-quality samples (e.g. with &gt;80% coverage).</p>
<p>A more serious issue affecting phylogenies is the presence of recurrent errors in certain positions of the genome. One of the regions with a higher prevalence of errors is the start and end of the consensus sequence, which also typically contains many missing data (see example in Figure 2). Therefore, it is common to <em>mask</em> the first and last few bases of the alignment, to avoid including spurious variable sites in the analysis.</p>
<div class="note">
<p>The term <strong>masking</strong> is often used to refer to the process of converting sequence bases to the ambiguous character ‘N’. You may come across this term in the documentation of certain tools, for example: “Positions with less than 20x depth of sequencing are masked.”</p>
<p>Masks are not limited to depth of sequencing. For example, <a href="https://ensemblgenomes.org/">reference genomes from ENSEMBL</a> are available with masked repeat or low-complexity sequences (e.g. around centromeres, transposon-rich regions, etc.).</p>
<p>The term <strong>soft masking</strong> is also used to refer to cases where, instead of using the ambiguous character ‘N’, sequences are masked with a lowercase. For example:</p>
<pre><code>&gt;seq_with_soft_masking
ACAGACTGACGCTGTcatgtatgtcgacGATAGGCTGATGGCGAGTGACTCGAG
&gt;seq_with_hard_masking
ACAGACTGACGCTGTNNNNNNNNNNNNNGATAGGCTGATGGCGAGTGACTCGAG</code></pre>
</div>
<p>Additionally, work by <a href="https://doi.org/10.1371/journal.pgen.1009175">Turakhia, de Maio, Thornlow, et al. (2020)</a> has identified several sites that show an unexpected mutation pattern. This includes, for example, mutations that unexpectedly occur multiple times in different parts of the tree (<a href="https://en.wikipedia.org/wiki/Homoplasy">homoplasies</a>) and often coincide with primer binding sites (from amplicon-based protocols) and can even be lab-specific (e.g. due to their protocols and data processing pipelines). The work from this team has led to the creation of a list of <a href="https://virological.org/t/masking-strategies-for-sars-cov-2-alignments/480">problematic sites</a>, which are recommended to be <em>masked</em> before running the phylogenetic analysis.</p>
<div class="figure">
<img src="https://journals.plos.org/plosgenetics/article/figure/image?size=inline&amp;id=10.1371/journal.pgen.1009175.g001" alt="" />
<p class="caption">Example of errors in phylogenetic inference due to recurrent sequencing errors. Source: <a href="https://journals.plos.org/plosgenetics/article/figure?id=10.1371/journal.pgen.1009175.g001">Figure 1 in Turakhia, de Maio, Thornlow et al. (2020)</a></p>
</div>
<p>So, let’s try to improve our alignment by masking the problematic sites, which are <a href="https://raw.githubusercontent.com/W-L/ProblematicSites_SARS-CoV2/master/problematic_sites_sarsCov2.vcf">provided as a VCF file</a>. This file also includes the first and last positions of the genome as targets for masking (positions 1–55 and 29804–29903, relative to the Wuhan-Hu-1 reference genome MN908947.3). The authors also provide a <a href="https://github.com/W-L/ProblematicSites_SARS-CoV2/blob/master/src/mask_alignment_using_vcf.py">python script</a> for masking a multiple sequence alignment. We have already downloaded these files to our course materials folder, so we can go ahead and use the script to mask our file:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/mask_alignment_using_vcf.py <span class="at">--mask</span> <span class="at">-v</span> resources/problematic_sites.vcf <span class="at">-i</span> results/mafft/uk_alignment.fa <span class="at">-o</span> results/mafft/uk_alignment_masked.fa</span></code></pre></div>
<p>If we open the output file with AliView, we can confirm that the positions specified in the VCF file have now been masked with the missing ‘N’ character.</p>
<p>We could then use this masked alignment for our tree-inference, just as we did before.</p>
<div class="note">
<p><strong>Using Python Scripts</strong></p>
<p>Bioinformaticians often write custom scripts for particular tasks. In this example, the authors of the “problematic sites” wrote a <em>Python</em> script that takes as input the FASTA file we want to mask as well as a VCF with the list of sites to be masked.</p>
<p>Python scripts are usually run with the <code>python</code> program and often accept options in a similar way to other command-line tools, using the syntax <code>--option</code> (this is not always the case, but most professionally written scripts follow this convention). To see how to use the script we can use the option <code>--help</code>. For our case, we could run:</p>
<pre class="console"><code>$ python scripts/mask_alignment_using_vcf.py --help</code></pre>
</div>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<div class="highlight">
<p><strong>Key Points</strong></p>
<ul>
<li>Methods for phylogenetic inference include <em>parsimony</em> and <em>maximum likelihood</em>. Maximum likelihood methods are preferred because they include more features of the evolutionary process. However, they are computationally more demanding than parsimony-based methods.</li>
<li>To build a phylogenetic tree we need a <em>multiple sequence alignment</em> of the sequences we want to infer a tree from.</li>
<li>In SARS-CoV-2, alignments are usually done against the Wuhan-Hu-1 reference genome.</li>
<li>We can use the software <code>mafft</code> to produce a multiple sequence alignment. The option <code>--addfragments</code> is used to produce an alignment against the reference genome.</li>
<li>The software <code>iqtree2</code> can be used for inferring trees from an alignment using maximum likelihood. This software supports a wide range of <em>substitution models</em> and a method to identify the model that maximizes the likelihood of the data.</li>
<li>Some of the substituion models that have been used to build global SARS-CoV-2 phylogenies are “GTR+G” and “GTR+I”.</li>
<li>We can time-scale trees using sample collection date information. The program <code>treetime</code> can be used to achieve this. For very large sample sizes (&gt;50,000 samples) the much faster program <em>Chronumental</em> can be used instead.</li>
<li>Before building a phylogeny, we should be careful to <em>mask</em> problematic sites that can lead to misleading placements of samples in the tree. The <a href="https://github.com/W-L/ProblematicSites_SARS-CoV2">SARS-CoV-2 Problematic Sites repository</a> provides with an updated list of sites that should be masked.</li>
</ul>
</div>
</div>
</div>

<!DOCTYPE html>
<html>
<head>
<style>
* {
  box-sizing: border-box;
}

.img-container-left {
  float: left;
  width: 33.33%;
  padding: 15px;
}

.img-container-right {
  float: right;
  width: 33.33%;
  padding: 15px;
}

.clearfix::after {
  content: "";
  clear: both;
  display: table;
}
</style>
</head>
<body>

<hr>
<div class="clearfix">
  <div class="img-container-left">
    <img src="assets/img/logo_btf.png" alt="UoC Bioinformatics Training Facility" style="width:100%">
  </div>
  <div class="img-container-right">
    <img src="assets/img/logo_ukhsa.svg" alt="UK Public Health England" style="width:40%">
  </div>
</div>

</body>
</html>





</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
