<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>SARS-CoV-2 Genomics</title>

<script src="site_libs/header-attrs-2.16/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="assets/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">UoC BioinfoTraining</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="101-setup.html">
    <span class="fa fa-gear"></span>
     
    Setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-file-lines"></span>
     
    Materials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">The Unix Shell</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="02a-unix_intro.html">Intro to Unix</a>
        </li>
        <li>
          <a href="02b-unix_files_directories.html">Files and Directories</a>
        </li>
        <li>
          <a href="02c-unix_text_manipulation.html">Text Manipulation</a>
        </li>
        <li>
          <a href="02d-unix_pipes.html">Combining Commands &amp; Writing Scripts</a>
        </li>
      </ul>
    </li>
    <li>
      <a href="01-intro.html">SARS-CoV-2 Genomic Surveillance</a>
    </li>
    <li>
      <a href="03-intro_ngs.html">Intro to NGS Data</a>
    </li>
    <li>
      <a href="04-consensus.html">Consensus Assembly</a>
    </li>
    <li>
      <a href="05-lineage_analysis.html">Lineages and Variants</a>
    </li>
    <li>
      <a href="06-phylogeny.html">Building Phylogenies</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Extras
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="102-unix_cheatsheet.html">Unix Cheatsheet</a>
    </li>
    <li>
      <a href="106-file_formats.html">Common File Formats</a>
    </li>
    <li>
      <a href="104-wsl_windows.html">Windows Subsystem for Linux</a>
    </li>
    <li>
      <a href="105-vs_code.html">Visual Studio Code</a>
    </li>
    <li>
      <a href="103-tools_and_resources.html">SARS-CoV-2 Resources</a>
    </li>
  </ul>
</li>
<li>
  <a href="100-homework.html">
    <span class="fa fa-pencil"></span>
     
    Homework
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cambiotraining/sars-cov-2-genomics">
    <span class="fa fa-brands fa-github"></span>
     
    Github
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<div id="extra-exercises" class="section level1">
<h1>Extra Exercises</h1>
<p>This set of exercises are provided as extra practice material that can be done outside of the course. It uses a larger set of data, which is more representative of what you may encounter in your own work.</p>
<p>All the exercises run from the data directory <code>homework</code>, so make sure you change into that directory. On our training machines this is <code>cd ~/Course_Materials/homework/</code>.</p>
<p>There are two sets of data, from two countries, one sequenced using Illumina and the other Nanopore platforms. Most of these exercises can be done on either dataset. We advise that you pick one of the datasets (for example, the one that uses data that you are most likely to work with) and go through all the exercises. After you complete them, you can start again with the other dataset.</p>
<div id="unix" class="section level2 tabset">
<h2 class="tabset">Unix</h2>
<p><strong>This exercise uses the skills covered in <a href="02a-unix_intro.html">Introduction to Unix</a>.</strong></p>
<ol style="list-style-type: decimal">
<li>Open a terminal and change into either the <code>homework/switzerland/</code> or <code>homework/southafrica/</code> directory.</li>
<li>Create a new directory called <code>results</code>.</li>
<li>Use command line tools to find how many sequencing read files are available (these are found within the <code>data/fastq_pass</code> or <code>data/reads/</code> directories).</li>
</ol>
<div id="south-africa" class="section level3">
<h3>South Africa</h3>
<div class="box">
<details>
<summary>
Answer
</summary>
<p>To check how many samples we have, we can combine the <code>ls</code> (list files) and <code>wc</code> (word count) commands</p>
<p>For the South Africa samples we have paired-end Illumina data (i.e. two files per sample). For this reason, we list all files that end with “_1.fastq.gz” using the <code>*</code> wildcard. This way we ensure to only count each sample once:</p>
<pre class="console"><code>$ ls data/reads/*_1.fastq.gz | wc -l</code></pre>
<pre><code>48</code></pre>
<p>We pipe the output of <code>ls</code> to the <code>wc</code> command using the option <code>-l</code> to count the number of lines coming out of the <code>ls</code> command.</p>
</details>
</div>
</div>
<div id="switzerland" class="section level3">
<h3>Switzerland</h3>
<div class="box">
<details>
<summary>
Answer
</summary>
<p>For the Switzerland samples we have Nanopore data. In this case, each sample’s data is within its own directory, so all we have to do is count how many directories there are in <code>data/fastq_pass</code>:</p>
<pre class="console"><code>$ ls data/fastq_pass/ | wc -l</code></pre>
<pre><code>59</code></pre>
<p>We pipe the output of <code>ls</code> to the <code>wc</code> command using the option <code>-l</code> to count the number of lines coming out of the <code>ls</code> command.</p>
</details>
</div>
</div>
</div>
<div id="metadata" class="section level2 tabset">
<h2 class="tabset">Metadata</h2>
<p><strong>This exercise uses the skills covered in <a href="01-intro.html">SARS-CoV-2 Genomic Surveillance</a>.</strong></p>
<p>Open the <code>homework</code> folder using the file browser <i class="fa-solid fa-folder"></i>.</p>
<ol style="list-style-type: decimal">
<li>Looking inside the each country’s folder, open the <code>README.txt</code> file to understand what kind of sequencing platform was used for these samples.</li>
<li>Open the metadata sheet (<code>sample_info.csv</code>) for one of the countries and look at the dates when these samples were collected.</li>
<li>Visit <a href="https://outbreak.info/location-reports">outbreak.info</a> and look at the location report for the country you chose. What variants might you expect to be find in these data?</li>
</ol>
<div id="south-africa-1" class="section level3">
<h3>South Africa</h3>
<div class="box">
<details>
<summary>
Answer
</summary>
<p>From the README we can see that using paired-end Illumina sequencing. This will need to be considered when we run our consensus assembly workflow.</p>
<p>Double-clicking the <code>sample_info.csv</code> file opens it with our spreadsheet software (on our training machines this is LibreOffice, but Excel would also open CSV files). The second column contains the sample collection dates, and we can see these samples are from Nov 2021 to Jan 2022.</p>
<p>We visit <a href="https://outbreak.info/location-reports">outbreak.info</a> and search for “South Africa”. If we scroll down to the section “Tracked lineages over time in South Africa” we can see that in late 2021 and the start of 2022 we have mostly the “Omicron” SARS-CoV-2 variant circulating in the population. This suggests that our samples may mostly be this variant.</p>
</details>
</div>
</div>
<div id="switzerland-1" class="section level3">
<h3>Switzerland</h3>
<div class="box">
<details>
<summary>
Answer
</summary>
<p>From the README we can see that these samples were sequenced using a Nanopore MinION platform. The raw Nanopore signal data was already processed to generate FASTQ files using Guppy. This will need to be considered when we run our consensus assembly workflow.</p>
<p>Double-clicking the <code>sample_info.csv</code> file opens it with our spreadsheet software (on our training machines this is LibreOffice, but Excel would also open CSV files). The second column contains the sample collection dates, and we can see these samples are from Nov/Dec 2021 to Jan 2022.</p>
<p>We visit <a href="https://outbreak.info/location-reports">outbreak.info</a> and search for “Switzerland”. If we scroll down to the section “Tracked lineages over time in Switzerland” we can see that around that period of 2021-2022 we had both Delta and Omicron variants circulating in the population, so we may expect to see these in our samples.</p>
</details>
</div>
</div>
</div>
<div id="fastq-quality-control" class="section level2 tabset">
<h2 class="tabset">FASTQ Quality Control</h2>
<p><strong>This exercise uses the skills covered in <a href="03-intro_ngs.html">Introduction to NGS</a>.</strong></p>
<div class="warning">
<p>This exercise can only be done on the South Africa samples.</p>
</div>
<p>The South Africa samples were sequenced on an Illumina platform. We will run a basic quality control of these reads using FastQC and compile the results using MultiQC.</p>
<ol style="list-style-type: decimal">
<li>Runs FastQC on all the samples. Output the results to a directory called <code>results/fastqc/</code>. Make sure to create the directory first.</li>
<li>Compile the results of FastQC using MultiQC and output its results in a directory called <code>results/multiqc/</code>.</li>
<li>Analyse the reports and make a note of any samples that you worry about in terms of producing a high-quality assembly.</li>
</ol>
<div class="box">
<details>
<summary>
Answer
</summary>
<p>To check the data quality of our reads, we can use the following commands. Make sure these commands are run from within the <code>homework/southafrica</code> directory.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># move to the directory</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/Course_Materials/homework/southafrica/</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># make directory</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/fastqc</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/multiqc</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># run FastQC on all the files</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co"># using 8 threads in parallel since we have 8 CPUs available</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-t</span> 8 <span class="at">--outdir</span> results/fastqc/ data/reads/<span class="pp">*</span>.fastq.gz</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co"># run MultiQC on the output of FastQC</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">--outdir</span> results/multiqc/ results/fastqc/</span></code></pre></div>
<p>This script starts by having some code to create the necessary output directories. The <code>-p</code> option ensures that we don’t get an error in case the directory already exists.</p>
<p>We then run <code>fastqc</code>, being careful to specify that we have 8 CPUs (with the <code>-t</code> option) to process the data in parallel. We also use the <code>*</code> <em>wildcard</em> to pattern-match all the files ending with the file extension “.fastq.gz”, so that FastQC will automatically process all the files.</p>
<p>We then use the <em>output</em> directory of the <code>fastqc</code> step as the <em>input</em> for <code>multiqc</code>. This is just like a mini bioinformatics workflow or our own: outputs of one tool feeding into the next tool!</p>
<p>After looking at the quality report from <code>multiqc</code>, we can notice that:</p>
<ul>
<li>Some samples have a very low number of sequences. For example, sample SRR17461700 only has 46717 reads (you can open the individual FastQC report for this sample to see the exact number).</li>
<li>Generally the read 2 files (file names ending <code>_2.fastq.gz</code>) have lower quality than the read 1 files. This is often seen in Illumina data.</li>
<li>In the “Adapter Content” section of the MultiQC report we can see that some of the reads contain some of the Illumina sequencing adapter within them.</li>
</ul>
<p>These last two points are generally solved with quality filtering and adapter removal, which is part of the consensus workflow we will use. However, it will still affect the quality of the assembly because by filtering low-quality reads we will, effectively be loosing data.</p>
<p>We should keep an eye on the samples with very low read numbers and check if they pass the QC thresholds from our consensus assembly workflow.</p>
</details>
</div>
</div>
<div id="consensus-assembly" class="section level2 tabset">
<h2 class="tabset">Consensus Assembly</h2>
<p><strong>This exercise uses the skills covered in <a href="04-consensus.html">Consensus Assembly</a>.</strong></p>
<p>We will now produce a consensus assembly from our sequencing data using the <code>nf-core/viralrecon</code> Nextflow pipeline. Consult the <a href="https://nf-co.re/viralrecon/2.4.1">full documentation</a> of this pipeline if you need to remind yourself what the different available options are.</p>
<ol style="list-style-type: decimal">
<li>Write a command to run our samples through the pipeline. (Note: this may take a couple of hours to complete.)
<ul>
<li>The input CSV file with sample names and FASTQ file locations has already been prepared and is called <code>samplesheet.csv</code>.</li>
<li>Output the results to a directory called <code>results/viralrecon/</code>.</li>
<li><strong>Bonus:</strong> Save the command used for your analysis in a script (create a <code>scripts/</code> folder and save it in there with a name of your choice).</li>
</ul></li>
<li>How long did the pipeline take to run, and did it give any warnings?</li>
<li>Open the <em>MultiQC</em> report file and explore the following questions:
<ul>
<li>What can you say about the quality of the consensus sequences in terms of the fraction of the genome that is covered? What do you think are reasons why some samples may be better than others?</li>
<li>Do you detect any issues with amplicon dropouts? Does this affect all samples or some samples specifically?</li>
<li>What types of mutation variants do you get, and are most of them silent or missence mutations?</li>
</ul></li>
</ol>
<details>
<summary>
Hint
</summary>
<p>Some of the files you may want to use are:</p>
<ul>
<li>The <em>MultiQC</em> report:
<ul>
<li>Illumina pipeline: <code>results/viralrecon/multiqc_report.html</code></li>
<li>Nanopore medaka pipeline: <code>results/viralrecon/medaka/multiqc_report.html</code></li>
</ul></li>
<li>The mutation variants file:
<ul>
<li>Illumina pipeline: <code>results/viralrecon/variants/ivar/variants_long_table.csv</code></li>
<li>Nanopore medaka pipeline: <code>results/viralrecon/medaka/variants_long_table.csv</code></li>
</ul></li>
<li>Open some of the BAM alignment files in IGV:
<ul>
<li>Illumina pipeline: <code>variants/bowtie2/*.ivar_trim.sorted.bam</code></li>
<li>Nanopore medaka pipeline: <code>results/viralrecon/medaka/*.primertrimmed.rg.sorted.bam</code></li>
</ul></li>
<li>The primer BED files can also be imported to IGV:
<ul>
<li>These are found in <code>resources/primers/artic_version3_pool1.bed</code> and <code>resources/primers/artic_version3_pool2.bed</code>.</li>
</ul></li>
</ul>
</details>
<div id="south-africa-2" class="section level3">
<h3>South Africa</h3>
<div class="box">
<details>
<summary>
Answer
</summary>
<p><strong>Question 1</strong></p>
<p>In order to run this pipeline, we need the following pieces of information:</p>
<ul>
<li>The platform used for sequencing (illumina or nanopore). This information is given to us in the <code>README.txt</code> file.</li>
<li>The primer set used and its version. This is also detailed in the <code>README.txt</code> file.</li>
<li>For Illumina data we also need an input CSV file with 3 columns: name of the sample, location of FASTQ read 1 file, location of FASTQ read 2 file. We have this information in the file <code>samplesheet.csv</code>.</li>
</ul>
<p>Our samples come from Illumina sequencing and used the ARTIC primer scheme version 3. Therefore, our command is:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run nf-core/viralrecon <span class="dt">\</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--input</span> samplesheet.csv <span class="dt">\</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outdir</span> results/viralrecon <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--protocol</span> amplicon <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--genome</span> <span class="st">&#39;MN908947.3&#39;</span> <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--primer_set</span> artic <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--primer_set_version</span> 3 <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--skip_assembly</span> <span class="dt">\</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--platform</span> illumina <span class="dt">\</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">-profile</span> singularity</span></code></pre></div>
<p>This step may take some hours to complete on our training machines, but we can see its progress printed on the console while it runs. Once it completes we should get a message printed on our screen, similar to:</p>
<pre><code>-[nf-core/viralrecon] 46/48 samples passed Bowtie2 1000 mapped read threshold:
    262615: ZA45
    225773: ZA32
    707111: ZA11
    292838: ZA35
    100358: ZA46
    2570: ZA43
    ..see pipeline reports for full list

-[nf-core/viralrecon] 2 samples skipped since they failed Bowtie2 1000 mapped read threshold:
    176: ZA12
    201: ZA09

-[nf-core/viralrecon] Pipeline completed successfully-
Waiting files transfer to complete (1 files)
Completed at: 04-May-2022 13:07:20
Duration    : 26m 56s
CPU hours   : 15.3
Succeeded   : 1&#39;901</code></pre>
<p>This tells us how long the pipeline took to run (in our example it was 26 minutes, but yours might be longer than this). We can also see a warning indicating that 2 samples were skipped due to an insufficient number of reads needed for downstream analysis. We will investigate this in the next question.</p>
<p><strong>Question 2</strong></p>
<p>After our pipeline completes, we can open the <em>MultiQC</em> report generated by the pipeline and located in <code>results/viralrecon/multiqc/multiqc_report.html</code>.</p>
<p>One of the first things we can look at is the summary table “Variant Calling Metrics” at the top of the report. We can start by looking at the two samples that were skipped in the analysis - ZA09 and ZA12 - and see that while they had a moderate number of starting reads (&gt;300 thousand), only around 200 reads were successfully mapped to the SARS-CoV-2 reference genome by <code>bowtie2</code>. Further looking at the table, we can see that this is because only a small percentage of reads - less than 2% - were non-host reads (i.e. non-human reads), which is determined by the <em>Kraken 2</em> software.</p>
<p>We can further explore this table to see that several other samples have very low % of non-human reads, suggesting several of these samples suffered from contamination of human DNA during preparation in the lab. We can see this in more detail in the section “PREPROCESS: Kraken 2”.</p>
<p>If we sort the table by the column “% Coverage &gt; 10x”, we can see that only 37 of the samples have this depth of coverage in more than 80% of the genome.</p>
<p>The “Amplicon coverage heatmap” also shows how several samples have low depth of coverage across most of the amplicons. There are also some amplicons that seem to have lower depth of coverage across multiple samples.</p>
<p>Finally, looking at the “Variants by Functional Class” section, we can disply the plot by as a percentage, which reveals most mutations are missense (i.e. they result in an amino acid change). Some samples have very low number of mutations (ZA10, ZA27, ZA30 and ZA31), making it difficult to draw conclusions from them.</p>
</details>
</div>
</div>
<div id="switzerland-2" class="section level3">
<h3>Switzerland</h3>
<div class="box">
<details>
<summary>
Answer
</summary>
<p><strong>Question 1</strong></p>
<p>In order to run this pipeline, we need the following pieces of information:</p>
<ul>
<li>The platform used for sequencing (illumina or nanopore). This information is given to us in the <code>README.txt</code> file.</li>
<li>The primer set used and its version. This is also detailed in the <code>README.txt</code> file.</li>
<li>For Nanopore data we also need an input CSV file with 2 columns: name of the sample, and the sample barcode number. We have this information in the file <code>samplesheet.csv</code>.</li>
<li>For basecalled data we also need to specify a model for the <code>medaka</code> software, in the format <code>{pore}_{device}_{caller variant}_{caller version}</code> (see <a href="https://github.com/nanoporetech/medaka#models">documentation here</a>). We are given this information in the <code>README.txt</code> file: R9.4.1 flowcell, MinION sequencer, high accuracy mode, Guppy version 3.6.0.</li>
</ul>
<p>Because we are starting with basecalled data (FASTQ files), we will use the pipeline with the <code>medaka</code> variant caller. Therefore, our command is:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run nf-core/viralrecon <span class="dt">\</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--input</span> samplesheet.csv <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outdir</span> results/viralrecon <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--protocol</span> amplicon <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--genome</span> <span class="st">&#39;MN908947.3&#39;</span> <span class="dt">\</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--primer_set</span> artic <span class="dt">\</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--primer_set_version</span> 3 <span class="dt">\</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--skip_assembly</span> <span class="dt">\</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--platform</span> nanopore <span class="dt">\</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--artic_minion_caller</span> medaka <span class="dt">\</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--artic_minion_medaka_model</span> r941_min_high_g360 <span class="dt">\</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">--fastq_dir</span> fastq_pass/ <span class="dt">\</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">-profile</span> singularity</span></code></pre></div>
<p>This step may take some hours to complete on our training machines, but we can see its progress printed on the console while it runs. Once it completes we should get a message printed on our screen, similar to:</p>
<pre><code>-[nf-core/viralrecon] Pipeline completed successfully-
Completed at: 04-May-2022 13:03:23
Duration    : 20m 2s
CPU hours   : 35.2
Succeeded   : 1&#39;312</code></pre>
<p>This tells us how long the pipeline took to run (in our example it was 20 minutes, but yours might be longer than this). We didn’t get any warnings, indicating that all samples were processed successfully.</p>
<p><strong>Question 2</strong></p>
<p>After our pipeline completes, we can open the <em>MultiQC</em> report generated by the pipeline and located in <code>results/viralrecon/multiqc/medaka/multiqc_report.html</code>.</p>
<p>One of the first things we can look at is the summary table “Variant Calling Metrics” at the top of the report. We can see that most samples had very good genome coverage at a depth &gt; 10%. If we sort the table by the column “% Coverage &gt; 10x”, we can see that the lowest sample had 75% of the genome covered at this depth, which is still a good fraction.</p>
<p>However, we can also note that two of the samples with the lowest genome coverage (CH07 and CH59) were not assigned a lineage by <em>Pangolin</em>, which suggests they did not pass the minimum thresholds for that program (we will explore more about lineage assignment in the next exercise). This may be due to a low number of reads, requiring re-sequencing these samples to achieve good genome coverage.</p>
<p>The “Amplicon coverage heatmap” shows an interesting pattern of amplicon dropout, with largely two groups of samples showing different patterns of amplicon depth of coverage. For example, amplicon “nCoV-2019_64” has extremely low depth of coverage in several samples such as CH20, CH43, CH24, CH40, CH44 amongst many others. Further investigation of this table reveals that these samples were classified as “Delta” by Nextclade. The other group of samples was classfied as “Omicron”, which may suggest that mutations in each SARS-CoV-2 variant are causing issues amplifying some of the PCR amplicons using the ARTIC v3 protocol.</p>
<p>A tip to explore these samples is to use the “highlight” button on MultiQC (on the right toolbar) and highlight certain samples of interest.</p>
<p>Further investigation of the amplicon dropout could be done by looking at the mutation file generated by the pipeline:</p>
<ul>
<li>Open the mutation variant file in a spreadhseet program - <code>results/viralrecon/medaka/variants_long_table.csv</code>.</li>
<li>Sort the table by position (“POS” column).</li>
<li>Check the location of the primers for amplicon “nCoV-2019_64”, for example from the <a href="https://github.com/artic-network/artic-ncov2019/blob/master/primer_schemes/nCoV-2019/V3/nCoV-2019.scheme.bed">BED file available online</a>. We can see that these primers are in positions 19204-19232 and 19591-19616.</li>
<li>Looking at the sorted table of variants we can see many samples have a SNP in position 19220, which overlaps the left primer. This could be a reason for this amplicon dropout.</li>
</ul>
<p>Finally, looking at the “Variants by Functional Class” section, we can disply the plot by as a percentage, which reveals most mutations are missense (i.e. they result in an amino acid change). Three samples have 1 nonsense mutation each (i.e. causing a new stop codon), which tend to be quite disruptive. Looking back at our mutation table, we can look at mutations with EFFECT column = “stop_gained” and will see that two of the samples share the same mutation (position 28209). This could be an indication that this is a true mutation rather than an error (since the chance of the same error occurring twice independently is low). However, this should be confirmed with new experiments as sequencing errors can sometimes be biased.</p>
</details>
</div>
</div>
</div>
<div id="cleaning-fasta-files" class="section level2 tabset">
<h2 class="tabset">Cleaning FASTA files</h2>
<p><strong>This exercise uses the skills covered in <a href="04-consensus.html#Cleaning_FASTA_Files">Consensus Assembly - Cleaning FASTA Files</a>.</strong></p>
<p>The <code>nf-core/viralrecon</code> pipeline generates a separate consensus sequence FASTA file for sample. However, for downstream analysis (variant analysis and phylogeny), it is more useful to have all the consensus sequecences combined into a single file.</p>
<ol style="list-style-type: decimal">
<li>Using command-line tools, combine all the FASTA files into a single file. Save the resulting file as <code>results/clean_consensus_sequences.fa</code>. (Note: see the hint below if you need a reminder of where these files are and what commands you can use.)</li>
<li>(Bonus - advanced) Using the <code>sed</code> command line tool, clean the names of the sequences such that only the original sample name is present in the final FASTA file. For example:
<ul>
<li>An illumina sample named as <code>&gt;ZA01 MN908947.3</code> should become <code>&gt;ZA01</code></li>
<li>A nanopore sample named as <code>&gt;CH01/ARTIC/medaka MN908947.3</code> should become <code>&gt;CH01</code></li>
</ul></li>
</ol>
<details>
<summary>
Hint
</summary>
<p>The FASTA files can be found in different directories, depending on whether you are processing the Nanopore or Illumina data:</p>
<ul>
<li>Illumina: fasta files are found in <code>results/viralrecon/variants/ivar/consensus/bcftools/</code> with <code>.fa</code> extension.</li>
<li>Nanopore: fasta files are found in <code>results/viralrecon/medaka/</code> with <code>.fasta</code> extension.</li>
</ul>
<p>Remember that all of the information about output files is <a href="https://nf-co.re/viralrecon/2.4.1/output">extensively detailed in the pipeline documentation</a>.</p>
<p>Some of the command-line tricks you can use to combine the files in this exercise include:</p>
<ul>
<li>Using the <code>cat</code> command to combine text files</li>
<li>Using the <code>*</code> wildcard to simultaneously select files matching a particular pattern in their name</li>
<li>Using the <code>&gt;</code> redirect operator to send the output of a command to a file (instead of printing it on the console).</li>
</ul>
</details>
<div id="south-africa-3" class="section level3">
<h3>South Africa</h3>
<div class="box">
<details>
<summary>
Answer
</summary>
<p><strong>Question 1</strong></p>
<p>For the South Africa samples (Illumina pipeline) the FASTA files can be found in the folder <code>results/viralrecon/variants/ivar/consensus/bcftools/</code>.</p>
<p>We can list all the files ending with <code>.fa</code> extension found in this folder:</p>
<pre class="console"><code>$ ls results/viralrecon/variants/ivar/consensus/bcftools/*.fa</code></pre>
<pre><code>results/viralrecon/variants/ivar/consensus/bcftools/ZA01.consensus.fa
results/viralrecon/variants/ivar/consensus/bcftools/ZA02.consensus.fa
results/viralrecon/variants/ivar/consensus/bcftools/ZA03.consensus.fa
results/viralrecon/variants/ivar/consensus/bcftools/ZA04.consensus.fa
results/viralrecon/variants/ivar/consensus/bcftools/ZA05.consensus.fa
[... more output not shown...]</code></pre>
<p>We can also count how many files we have and confirm this matches the number of samples in this dataset:</p>
<pre class="console"><code>$ ls results/viralrecon/variants/ivar/consensus/bcftools/*.fa | wc -l</code></pre>
<pre><code>45</code></pre>
<p>Finally, we can combine all the files using the <code>cat</code> command and redirecting the output with <code>&gt;</code> to a new file:</p>
<pre class="console"><code>$ cat results/viralrecon/variants/ivar/consensus/bcftools/*.fa &gt; results/clean_consensus_sequences.fa</code></pre>
<p>We can confirm that all the samples are present in this file:</p>
<pre class="console"><code>$ cat results/clean_consensus_sequences.fa | grep &quot;&gt;&quot; | wc -l</code></pre>
<pre><code>45</code></pre>
<ul>
<li>We print the content of the file with <code>cat</code>.</li>
<li>We find lines of the file that contain the <code>&gt;</code> character (the sequence names in FASTA files always start with this character).</li>
<li>We use the <code>wc -l</code> to count the number of lines in the output.</li>
</ul>
<p><strong>Question 2</strong></p>
<p>Looking at the content of one of the files using <code>head</code>, we can see how the sequences are named:</p>
<pre class="console"><code>$ head -n 3 results/viralrecon/variants/ivar/consensus/bcftools/ZA01.consensus.fa</code></pre>
<pre><code>&gt;ZA01 MN908947.3
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNAGATCT
GTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACT</code></pre>
<p>We can remove the ” MN908947.3” text from the sample names using the <code>sed</code> text-replacement function. This would be our modified command to combine all samples and remove that text:</p>
<pre class="console"><code>$ cat results/viralrecon/medaka/*.fasta | sed &#39;s/ MN908947.3//&#39; &gt; results/clean_consensus_sequences.fa</code></pre>
<p>Some notes about the <code>sed</code> command:</p>
<ul>
<li>The syntax to substitute text is <code>sed 's/old text/new text/'</code>.</li>
<li>If we want to <em>remove</em> text we can leave the new text field empty: <code>sed 's/old text//'</code></li>
<li>Notice that we want to also substitute the <em>space</em> before “MN908947.3”, so we include that in our pattern to substitute.</li>
</ul>
</details>
</div>
</div>
<div id="switzerland-3" class="section level3">
<h3>Switzerland</h3>
<div class="box">
<details>
<summary>
Answer
</summary>
<p><strong>Question 1</strong></p>
<p>For the Switzerland samples (Nanopore <code>medaka</code> pipeline) the FASTA files can be found in the folder <code>results/viralrecon/medaka/</code>.</p>
<p>We can list all the files ending with <code>.fasta</code> extension found in this folder:</p>
<pre class="console"><code>$ ls results/viralrecon/medaka/*.fasta</code></pre>
<pre><code>results/viralrecon/medaka/CH01.consensus.fasta
results/viralrecon/medaka/CH03.consensus.fasta
results/viralrecon/medaka/CH05.consensus.fasta
results/viralrecon/medaka/CH06.consensus.fasta
results/viralrecon/medaka/CH07.consensus.fasta
[... more output not shown...]</code></pre>
<p>We can also count how many files we have and confirm this matches the number of samples in this dataset:</p>
<pre class="console"><code>$ ls results/viralrecon/medaka/*.fasta | wc -l</code></pre>
<pre><code>59</code></pre>
<p>Finally, we can combine all the files using the <code>cat</code> command and redirecting the output with <code>&gt;</code> to a new file:</p>
<pre class="console"><code>$ cat results/viralrecon/medaka/*.fasta &gt; results/clean_consensus_sequences.fa</code></pre>
<p>We can confirm that all the samples are present in this file:</p>
<pre class="console"><code>$ cat results/clean_consensus_sequences.fa | grep &quot;&gt;&quot; | wc -l</code></pre>
<pre><code>59</code></pre>
<ul>
<li>We print the content of the file with <code>cat</code>.</li>
<li>We find lines of the file that contain the <code>&gt;</code> character (the sequence names in FASTA files always start with this character).</li>
<li>We use the <code>wc -l</code> to count the number of lines in the output.</li>
</ul>
<p><strong>Question 2</strong></p>
<p>Looking at the content of one of the files using <code>head</code>, we can see how the sequences are named:</p>
<pre class="console"><code>$ head -n 3 results/viralrecon/medaka/CH01.consensus.fasta</code></pre>
<pre><code>&gt;CH01/ARTIC/medaka MN908947.3
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNAGATCT
GTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACT</code></pre>
<p>We can remove the “/ARTIC/medaka MN908947.3” text from the sample names using the <code>sed</code> text-replacement function. This would be our modified command to combine all samples and remove that text:</p>
<pre><code>cat results/viralrecon/medaka/*.fasta | sed &#39;s/\/ARTIC\/medaka MN908947.3//&#39; &gt; results/clean_consensus_sequences.fa</code></pre>
<p>Some notes about the <code>sed</code> command:</p>
<ul>
<li>The syntax to substitute text is `sed ‘s/old text/new text/’.</li>
<li>If we want to <em>remove</em> text we can leave the new text field empty: <code>sed 's/old text//'</code></li>
<li>Because <code>/</code> is used to separate the different parts of the <code>sed</code> command, when we want to replace the actual “/” letter we need to prefix it <code>\</code> (this is called “escaping”). Therefore, we use “/ARTIC/medaka MN908947.3” as the text to replace.</li>
</ul>
</details>
</div>
</div>
</div>
<div id="lineages-variants" class="section level2 tabset">
<h2 class="tabset">Lineages &amp; Variants</h2>
<p>We will now investigate in more detail the SARS-CoV-2 lineages/clades and variants of concern in our samples.</p>
<ol style="list-style-type: decimal">
<li>Looking at the <em>MultiQC</em> report, answer the following:
<ul>
<li>How many different variants of concern were detected in your samples?</li>
<li>Looking at the <em>Pangolin</em> analysis section, where there any problematic samples and, if so, what was the reason?</li>
</ul></li>
<li>Run a full <em>Nextclade</em> analysis using the web application: <a href="https://clades.nextstrain.org/" class="uri">https://clades.nextstrain.org/</a>. Upload the clean FASTA file you generated in the previous exercise.
<ul>
<li>What quality issues do you detect using this tool? Does it agree with your previous conclusions?</li>
</ul></li>
</ol>
<div id="south-africa-4" class="section level3">
<h3>South Africa</h3>
:::box
<details>
<summary>
Answer
</summary>
<p><strong>Question 1</strong></p>
<p>There are two main tools used to assign consensus sequences to lineages/clades and classify them as variants of concern: <em>Nextclade</em> and <em>Pangolin</em>.</p>
<p>We can look at the first table of the <em>MultiQC</em> report and look at the column “Nextclade clade” to see which SARS-CoV-2 variants were detected using this tool. We can see that all samples were classified as Omicron, except one sample, which was not classified as a known variant of concern: ZA10 was assigned to <em>Nextclade</em> clade 20C.</p>
<p>Looking at the section of the <em>MultiQC</em> report called “VARIANTS: Pangolin” we can see what variants were detected with this tool (which uses <code>scorpio</code> to do the classification of variants of concern). Again, the analysis for this tool agrees with the <em>Nextclade</em> analysis, in that all samples were classified as Omicron.</p>
<p>However, <em>Pangolin</em> seems to have a higher number of samples that were not classified, as they did not pass the minimum genome coverage threshold required for its analysis. These are highlighted as “Fail” in the column “QC Status”. In the column “Note” we can see the reason was a high proportion of missing data (“N”).</p>
<p><strong>Question 2</strong></p>
<p>We go to <a href="https://clades.nextstrain.org/" class="uri">https://clades.nextstrain.org/</a> and:</p>
<ul>
<li>Click <strong>Select a file</strong> to browse your computer and upload the FASTA file with the cleaned consensus sequences (<code>results/clean_consensus_sequences.fa</code>).</li>
<li>Click <strong>Run</strong></li>
</ul>
<p>We are then presented with the <em>Nextclade</em> interactive results panel, similar to the one shown in this picture:</p>
<p><img src="images/nextclade_overview.svg" /></p>
<p>We can sort the table by the column “QC” (quality control) to look at the most problematic samples. We can see the main reason for low QC score is a high proportion of missing data, which is in agreement with the information we previously looked at from <em>Pangolin</em>.</p>
<p>However, other issues are also present in some of the samples:</p>
<ul>
<li>The occurrence of private mutations, which are mutations that are not present in any other sample and so could be due to sequencing errors.</li>
<li>the occurrence of frame shifts, which are insertion/deletion mutations that change the translation frame of a protein. Such a mutation would be very disruptive and likely lethal, so it may also be caused by sequencing errors.</li>
</ul>
</details>
</div>
<div id="switzerland-4" class="section level3">
<h3>Switzerland</h3>
:::box
<details>
<summary>
Answer
</summary>
<p><strong>Question 1</strong></p>
<p>There are two main tools used to assign consensus sequences to lineages/clades and classify them as variants of concern: <em>Nextclade</em> and <em>Pangolin</em>.</p>
<p>We can look at the first table of the <em>MultiQC</em> report and look at the column “Nextclade clade” to see which SARS-CoV-2 variants were detected using this tool. We can see that samples were either classified as Omicron or Delta.</p>
<p>Looking at the section of the <em>MultiQC</em> report called “Pangolin” we can see what variants were detected with this tool (which uses <code>scorpio</code> to do the classification of variants of concern):</p>
<ul>
<li>The analysis for this tool agrees with the <em>Nextclade</em> analysis, in that samples were classified either as Omicron or Delta.</li>
<li>However, some samples (CH22, CH47, CH58 and CH62) were classified as “Probable Omicron”, suggesting that there were not enough known SNPs from these SARS-CoV-2 variants present in the consensus sequences. This is indicated by the column “S support”, which has a lower score for these samples.</li>
<li>Two samples (CH07 and CH59) were not classified. These are highlighted as “Fail” in the column “QC Status”. In the column “Note” we can see the reason was a high proportion of missing data (“N”).</li>
</ul>
<p><strong>Question 2</strong></p>
<p>We go to <a href="https://clades.nextstrain.org/" class="uri">https://clades.nextstrain.org/</a> and:</p>
<ul>
<li>Click <strong>Select a file</strong> to browse your computer and upload the FASTA file with the cleaned consensus sequences (<code>results/clean_consensus_sequences.fa</code>).</li>
<li>Click <strong>Run</strong></li>
</ul>
<p>We are then presented with the <em>Nextclade</em> interactive results panel, similar to the one shown in this picture:</p>
<p><img src="images/nextclade_overview.svg" /></p>
<p>We can sort the table by the column “QC” (quality control) to look at the most problematic samples. We can see the main reason for low QC score is a high proportion of missing data affecting many of the samples.</p>
<p>Two samples (CH38 and CH26) have also been highlighted as having too many private mutations, which are mutations that are not present in any other sample and so could be due to sequencing errors.</p>
</details>
<!--
## Phylogenetics {.tabset}

### South Africa

:::box
<details><summary>Answer</summary>

</details>

### Switzerland

:::box
<details><summary>Answer</summary>

</details>
-->
</div>
</div>
</div>

<!DOCTYPE html>
<html>
<head>
<style>
* {
  box-sizing: border-box;
}

.img-container-left {
  float: left;
  width: 33.33%;
  padding: 15px;
}

.img-container-right {
  float: right;
  width: 33.33%;
  padding: 15px;
}

.clearfix::after {
  content: "";
  clear: both;
  display: table;
}
</style>
</head>
<body>

<hr>
<div class="clearfix">
  <div class="img-container-left">
    <img src="assets/img/logo_btf.png" alt="UoC Bioinformatics Training Facility" style="width:100%">
  </div>
  <div class="img-container-right">
    <img src="assets/img/logo_ukhsa.svg" alt="UK Public Health England" style="width:40%">
  </div>
</div>

</body>
</html>





</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
